<!--
--- UCSF Chimera Copyright ---
Copyright (c) 2005-2009 Regents of the University of California.
All rights reserved.  This software provided pursuant to a
license agreement containing restrictions on its disclosure,
duplication and use.  This notice must be embedded in or
attached to all copies, including partial copies, of the
software or any revisions or derivations thereof.
--- UCSF Chimera Copyright ---
-->

<html><head>
<title>Attributes Tutorial</title>
</head><body>
<font size="-1">
<a href="../frametut.html" target="_top">Chimera Tutorials Index</a>
</font>

<h2>Attributes Tutorial</h2>
<p>
This tutorial demonstrates many uses of
<a href="../../ContributedSoftware/defineattrib/defineattrib.html#attribdef">
attributes</a>, or properties of atoms, residues, and molecule models.
Attributes can be numerical (such as atomic number),
boolean (<i>e.g.</i>, whether a residue is in a helix), 
or string-valued (such as <a href="../idatm.html">atom type</a>).
Attribute values can be rendered visually
and used in <a href="../../UsersGuide/selection.html">selection</a>
and command-line
<a href="../../UsersGuide/midas/frameatom_spec.html" target="_top">atom 
specification</a>.
</p><p>
<a href="#part1">Part 1</a> uses a leucine zipper structure, and
<a href="#part2">Part 2</a> uses the structure of a GTP-binding protein.
</p>
<hr>
<a name="part1">
<h3>Attributes, Part 1 - Leucine Zipper</h3></a>

<table align="right" style="margin:8px 8px 8px 8px">
<tr><th align="center">
<a name="fig1">tyrosine colored by charge</a></th></tr>
<tr><td><img src="att1.png"></td></tr>
</table>

<p>
On <b>Windows</b>/<b>Mac</b>, click the <b>chimera</b> icon;
on <b>UNIX</b>, start Chimera from the system prompt:
<blockquote><tt>
<i>unix</i>: <b>chimera</b>
</tt></blockquote>
A splash screen will appear, to be replaced in a few seconds
by the main Chimera window.  Show the
<a href="../chimerawindow.html#emulator"><b>Command Line</b></a>
(choosing <b>Tools... General Controls... Command Line</b> is one way)
and optionally, the
<a href="../sideview.html#sideview"><b>Side View</b></a>  
for clipping and scaling.
</p><p>
If you have internet connectivity, structures can be obtained directly from the
<a href="http://www.rcsb.org" target="_blank">Protein Data Bank</a>.
Choose <b>File... <a href="../fetch.html">Fetch by ID</a></b>
from the Chimera menu and use the resulting dialog to
fetch <b>1zik</b> from the <b>PDB</b>.
If you do not have internet connectivity,
<a href="download.html">download</a> the file <a href="1zik.pdb">1zik.pdb</a>
included with this tutorial and use <b>File... Open</b> to open it.
</p><p>
The structure is a leucine zipper formed by two peptides.
Apply the "all atoms" <a href="../menu.html#menupresets">preset</a>
and undisplay water:
<blockquote>
<i>Command</i>: <b><a href="../midas/preset.html">preset</a> apply interactive 2</b>
<br><i>Command</i>: <b><a href="../midas/display.html">~disp</a> solvent</b>
</blockquote>
<a href="../mouse.html">Move and scale</a> the structure
as desired throughout the tutorial.  
<p><p>
Open the <a href="../../ContributedSoftware/render/render.html#render">
<b>Render by Attribute</b></a> tool (<b>Tools... Structure Analysis... 
Render by Attribute</b>).
Make sure it is set to show the attributes of <b>atoms</b>.  Look in the
<b>Attribute</b> list to see the attributes available: <b>bfactor</b>
and <b>occupancy</b>, which were read from the input PDB file.
Choose <b>bfactor</b>; a histogram of the values will
appear, with blue, white and red markers (or 
<a href="../../ContributedSoftware/render/render.html#thresh">thresholds</a>).
Click <b>Apply</b> to color the atoms blue to white to red in 
order of increasing B-factor.
</p><p>
You can define your own color mapping by moving thresholds 
along the histogram and/or changing their colors.
The <b>Value</b> and <b>Color</b> are shown for the
most recently clicked or moved threshold.  The <b>Value</b> changes
when the threshold is moved, or the position can be changed by
entering a value and pressing return. The <b>Color</b> can be changed
by clicking the square <a href="../coloring.html#colorwell">color well</a>
and using the <a href="../colortool.html"><b>Color Editor</b></a>.
Thresholds can be added or deleted by Ctrl-clicking on the histogram.
If you wish, try applying different color mappings.  
</p><p>
As expected, the
atoms with higher B-factors are on the outside of the structure.
Note that the histogram includes the B-factor values of the waters
even though they are not displayed.
Display waters with B-factors less than 75:
<blockquote>
<i>Command</i>: <b><a href="../midas/display.html">disp</a> solvent & @/bfactor&lt;75</b>
<br><i>Command</i>: <b><a href="../midas/represent.html">repr bs</a> solvent</b>
</blockquote>
Show ribbons:
<blockquote>
<i>Command</i>: <b><a href="../midas/ribbon.html">ribbon</a></b>
</blockquote>
Next, change to the attributes of <b>residues</b>
in the <b>Render by Attribute</b> dialog.  Automatically available residue
attributes include <b>kdHydrophobicity</b> and average <b>bfactor</b>
and <b>occupancy</b>.  The latter two are residue averages over the
atomic values.  Choose <b>kdHydrophobicity</b>, the
<a href="../midas/hydrophob.html">Kyte-Doolittle hydrophobicity scale</a> 
for amino acids.  The values will be displayed in the histogram:
negative for polar residues, positive for hydrophobic residues. 
</p><p>
Hydrophobicity could be shown with color, but for variety,
we will show it with "worms," modified ribbons that vary in radius.
In the <b>Render by Attribute</b> dialog, change from <b>Colors</b>
to <b>Worms</b>.
The values are still shown in a histogram, but now the thresholds have
a <b>Worm radius</b> instead of a color.  The thresholds can be added,
deleted, and moved like before, and the <b>Worm radius</b> can be changed.
By default, the more hydrophobic residues will be shown with a fatter worm
(a larger worm radius).  Change the mapping as desired and click <b>Apply</b>.
</p><p>
The worms show that the most hydrophobic residues 
tend to face the interior of the structure. To return to a normal ribbon
instead of a worm, change the <b>Worm style</b> to <b>non-worm</b>
and click <b>OK</b> (which is equivalent to <b>Apply</b> plus <b>Close</b>).
</p><p>
Additional <a href="../midas/hydrophob.html">hydrophobicity scales</a>
are available as
<a href="../../ContributedSoftware/defineattrib/defineattrib.html#attrfile">attribute
assignment files</a>.
</p><p>
Some Chimera tools create new attributes.  For example, the
<a href="../../ContributedSoftware/addcharge/addcharge.html"><b>Add 
Charge</b></a> tool (also implemented as the command
<b><a href="../midas/addcharge.html">addcharge</a></b>)
assigns partial charges as an atom attribute named <b>charge</b>.
Charge values from
<a href="http://ambermd.org/index.html" target="_blank"><b>Amber</b></a> 
are used for standard amino acids and nucleotides.
</p><p>
Use the "all atoms" preset again:
<blockquote>
<i>Command</i>: <b><a href="../midas/preset.html">preset</a> apply int 2</b>
</blockquote>
Next, add hydrogens and assign atomic partial charges: 
<blockquote>
<i>Command</i>: <b><a href="../midas/addh.html">addh</a></b>
<br><i>Command</i>: <b><a href="../midas/addcharge.html">addcharge</a></b>
</blockquote>
Above, attribute values were shown with color and worms; another approach
is to display the values as atom labels.  Since it would be too confusing to
show all of these labels at once, only show them for the atoms in a
single residue: 
<blockquote>
<i>Command</i>: <b><a href="../midas/alias.html">alias</a> myres :tyr.a</b>
<br><i>Command</i>: <b><a href="../midas/show.html">show</a> myres</b>
<br><i>Command</i>: <b><a href="../midas/focus.html">focus</a></b>
<br><i>Command</i>: <b><a href="../midas/labelopt.html">labelopt info</a> charge</b>
<br><i>Command</i>: <b><a href="../midas/label.html">label</a> myres</b>
</blockquote>
</p><p>
New attributes can be used in the command line just like
built-in attributes.  For example, show the atoms with <b>charge</b>
less than &ndash;0.4 as balls:
<blockquote>
<i>Command</i>: <b><a href="../midas/represent.html">repr bs</a> @/charge&lt;-0.4</b>
</blockquote>
Finally, remove the labels, change to ball-and-stick, and
color atoms by their <b>charge</b> values:
<blockquote>
<i>Command</i>: <b><a href="../midas/label.html">~la</a></b>
<br><i>Command</i>: <b><a href="../midas/represent.html">repr bs</a></b>
<br><i>Command</i>: <b><a href="../midas/rangecolor.html">rangecol</a> charge -0.8 red 0 white 0.8 blue</b>
</blockquote>
The result should look something like the <a href="#fig1">figure</a>.
<a href="../midas/rangecolor.html"><b>Rangecolor</b></a>
is the command alternative to <b>Render by Attribute</b> for coloring.
</p><p>
Close the model:
<blockquote>
<i>Command</i>: <b><a href="../midas/close.html">close</a> 0</b>
</blockquote>
Go on to <a href="#part2"><b>Part 2</b></a> below, <b>OR</b>
terminate the Chimera session with the following command:
<blockquote>
<i>Command</i>: <b><a href="../midas/stop.html">stop</a></b>
</blockquote>
</p>
<hr>
<a name="part2">
<h3>Attributes, Part 2 - GTP-Binding Protein</h3></a>

<table align="right" style="margin:8px 8px 8px 8px">
<tr><th align="center">
<a name="fig2">121p colored by convexity</a></th></tr>
<tr><td><img src="att2.png"></td></tr>
</table>

<p>
Begin with Chimera started and the
<a href="../chimerawindow.html#emulator"><b>Command Line</b></a>
(and optionally the <a href="../sideview.html#sideview"><b>Side View</b></a>)
opened as described at the beginning of <a href="#part1"><b>Part 1</b></a>.
</p><p>
If you have internet connectivity,
use a command to fetch the PDB structure <b>121p</b>:
<blockquote>
<i>Command</i>: <b><a href="../midas/open.html">open</a> 121p</b>
</blockquote>
If you do not have internet connectivity,
<a href="download.html">download</a> the file <a href="121p.pdb">121p.pdb</a> 
included with this tutorial and use <b>File... Open</b> to open it.
</p><p>
The structure is H-ras, a small GTP-binding protein, along with
a bound GTP analog, a Mg<sup>++</sup> ion, and some water.
<a href="../mouse.html">Move and scale</a> the structure
as desired throughout the tutorial.
</p><p>
Use the ribbons <a href="../menu.html#menupresets">preset</a>: 
<blockquote>
<i>Command</i>: 
<b><a href="../midas/preset.html">preset</a> apply int 1</b>
</blockquote>
This shows a ribbon rainbow-color-coded from blue at the N-terminus 
to red at the C-terminus, plus atoms in or near binding sites.
Delete the water (most of which is hidden)
and label the residues near the ion:
<blockquote>
<i>Command</i>: <b><a href="../midas/delete.html">delete</a> solvent</b>
<br>
<i>Command</i>: <b><a href="../midas/rlabel.html">rlabel</a> ions z<3.5</b>
</blockquote>
The active site Mg<sup>++</sup> ion is coordinated by Ser 17, Thr 35, 
and phosphonate oxygens in the GTP analog, which is named GCP.
Color the ligand GCP yellow and remove the residue labels:
<blockquote>
<i>Command</i>: <b><a href="../midas/color.html">color</a> yellow ligand</b>
<br><i>Command</i>: <b><a href="../midas/rlabel.html">~rlabel</a></b>
</blockquote>
In <a href="http://tardis.nibio.go.jp/homstrad/"
target="_blank">HOMSTRAD</a>, this protein is classified in the 
<a href="http://tardis.nibio.go.jp/cgi-bin/homstrad/showpage.cgi?family=gtp&disp=str" 
target="_blank">GTP-binding protein</a> family.
<a href="download.html">Download</a> the alignment file
<a href="homstrad-gtp.pir">homstrad-gtp.pir</a>
(originally from HOMSTRAD, now included with this tutorial)
and open it with <b>File... Open</b>.  The file type is
<b>Aligned NBRF/PIR</b>.  The alignment will be shown with
<a href="../../ContributedSoftware/multalignviewer/framemav.html"
target="_top"><b>Multalign Viewer</b></a>.  The structure will automatically
associate with the sequence named <b>5p21</b>, as indicated by the white
rectangle under this sequence name.
</p><p>
When a structure is associated with a sequence alignment shown by
<a href="../../ContributedSoftware/multalignviewer/framemav.html" 
target="_top"><b>Multalign Viewer</b></a>, its residues are assigned
<a href="../../ContributedSoftware/multalignviewer/multalignviewer.html#mavAttributes">
conservation attributes</a>.  From the <b>Multalign Viewer</b>
menu, choose <b>Structure... Render by Conservation</b>.  This opens the same
<a href="../../ContributedSoftware/render/render.html#render"><b>Render 
by Attribute</b></a> tool used in <a href="#part1"><b>Part 1</b></a>, except
that now it shows the attribute of <b>residues</b> named 
<b>mavConservation</b>. The "mav" part of the name is shorthand for
<b>Multalign Viewer</b>, and "Conservation" indicates that the values
correspond to what is shown in the <b>Conservation</b> line of the
<a href="../../ContributedSoftware/multalignviewer/multalignviewer.html#seqwindow">sequence window</a>.
By default, the value for a column is the fraction of sequences with
the most common residue type at that position, for example:
0.7 where 7 of 10 sequences have the same type of residue,
0.1 where each of the 10 has a different type.
(Different calculation methods are available, see the <b>Multalign Viewer</b>
<a href="../../ContributedSoftware/multalignviewer/multalignviewer.html#mavpref-analysis"><b>Headers</b> preferences</a>.)
</p><p>
In <a href="../../ContributedSoftware/render/render.html#render"><b>Render
by Attribute</b></a>, use <b>Colors</b> and
adjust the color mapping as desired before clicking <b>Apply</b>.  
The coloring shows that many positions in or near the binding site
are highly conserved.
</p><p>
The <b>Render by Attribute</b> dialog also lists the residue attribute
<b>mavPercentConserved</b>.  Whereas <b>mavConservation</b> changes
along with the <b>Conservation</b> calculation method, 
<b>mavPercentConserved</b> is always the percentage of sequences 
with the most common residue type at that position
(same as the <b>Conservation</b> default, except expressed
as a percentage rather than a fraction).
</p><p>
Select residues within 6 &Aring; of the ligand:
<blockquote>
<i>Command</i>: <b><a href="../midas/select.html#newer">select</a> ligand z&lt;6</b>
</blockquote>
The selection is shown with green boxes on the sequence alignment.
</p><p>
Like other attributes, <b>mavConservation</b> and <b>mavPercentConserved</b>
can be used in commands:
<blockquote>
<i>Command</i>: <b><a href="../midas/select.html#newer">select</a> ligand z&lt;6 & :/mavPercentConserved&gt;80</b>
</blockquote>
This selects a smaller set of residues than before, only those both
within 6 &Aring; of the ligand and meeting the conservation criterion.
Clear the selection (<b>Select... Clear Selection</b> is one way) and
<b>Quit</b> from <b>Multalign Viewer</b>.
</p><p>
Show the molecular surface of the protein:
<blockquote>
<i>Command</i>: <b><a href="../midas/surface.html">surface</a></b>
</blockquote>
By default, the surface colors match the underlying atom colors. It is again 
evident that highly conserved residues line the binding pocket.
Generating a molecular surface automatically creates the atom attributes
<b>areaSAS</b> and <b>areaSES</b>, solvent-accessible
and solvent-excluded surface areas in the context of the structure.  
For example, the values of both attributes are zero for
atoms buried in the protein interior.  What Chimera shows is 
the <b><i>solvent-excluded</i></b> molecular surface,
composed of probe contact, toroidal, and reentrant surface.
The <b><i>solvent-accessible</i></b> surface (not displayed)
is farther out from the molecule, the surface traced out by the probe center.
</p><p>
If <b>Render by Attribute</b> is not already open, open it 
(<b>Tools... Depiction... Render by Attribute</b>) and see 
what attributes are present for <b>atoms</b>.  
If <b>areaSAS</b> and <b>areaSES</b> are not yet listed,
show them with <b>Refresh... Menus</b>.
The histograms show the ranges of values obtained for the structure.
</p><p>
Open the
<a href="../../ContributedSoftware/calculator/calculator.html"><b>Attribute
Calculator</b></a> (<b>Tools... Structure Analysis... Attribute Calculator</b>).
Calculate a new attribute named <b>convexity</b> for <b>atoms</b> 
using the <b>Formula</b>
<blockquote>
atom.areaSAS/atom.areaSES
</blockquote>
Click <b>OK</b> to perform the calculation and assignment.  
A warning message will appear because some atoms have an
<b>areaSES</b> of zero, resulting in a divide-by-zero error. 
However, just close the warning dialog; the attribute 
has been assigned correctly for the atoms with nonzero areas.
</p><p>
Finally,
show the <b>convexity</b> values with color on the protein surface.
In the <b>Render by Attribute</b> dialog, 
make sure that the histogram is showing the
new attribute of <b>atoms</b> named <b>convexity</b>.  In the
<b>Colors</b> section, change the colors by clicking a threshold,
clicking the <a href="../coloring.html#colorwell">color well</a>, and then
entering a <b>Color name</b> in the <a href="../colortool.html">
<b>Color Editor</b></a>: use <b>purple</b> for the lowest-value (leftmost)
threshold and <b>yellow</b> for the highest-value threshold.
Use <b>white</b> for the middle threshold and set its <b>Value</b> to 1
(values &gt; 1 represent convex areas, while values &lt; 1 represent
concave areas).  Click <b>Apply</b>;
the result should resemble the <a href="#fig2">figure</a>.
</p><p>
When finished, end the Chimera session:
<blockquote>
<i>Command</i>: <b><a href="../midas/stop.html">stop really</a></b>
</blockquote>
</p>
<hr>
<address>meng-at-cgl.ucsf.edu / August 2009</address>
</body></html>
