<!doctype html public "-//W3C//DTD HTML 4.0//EN">
<!-- vim:sw=2:
>>> UCSF Chimera Copyright
Copyright (c) 2008-2009 Regents of the University of California.
All rights reserved.  This software provided pursuant to a
license agreement containing restrictions on its disclosure,
duplication and use.  This notice must be embedded in or
attached to all copies, including partial copies, of the
software or any revisions or derivations thereof.
<<< UCSF Chimera Copyright
-->
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=ISO-8859-1">
<link rel="icon" href="chimera16.png" type="image/png">
<title>Web Services Used by UCSF Chimera</title>
</head>
<body>
<p>
<img src="titleChimera.png" width="370" height="55" alt="logo">
</p>

<h2>Web Services Used by UCSF Chimera</h2>

<p>
Chimera connects to the following sites for data retrieval and calculations.
See also:
<a href="#webdatasites">sites that provide Chimera Web data</a>
</p>

Atomic coordinates:
<blockquote>

<dt><a name="pdb" href="http://www.rcsb.org" 
target="_blank"><b>Protein Data Bank</b></a> (PDB)
<dd>
Allows <a href="UsersGuide/fetch.html">fetching</a> structures 
(in PDB and mmCIF formats) by
<a href="http://www.rcsb.org/robohelp_f/search_database/pdb_identifier.htm"
target="_blank">PDB ID</a>.
<br><br>

<dt><a name="ndb" <a href="http://ndbserver.rutgers.edu/index.html" 
target="_blank"><b>Nucleic Acid Database</b></a> (NDB)
<dd>
Translates NDB ID codes into PDB ID codes, which are then used to 
<a href="UsersGuide/fetch.html">fetch</a> structures
from the <a href="#pdb">PDB</a>.
<br><br>

<dt><a name="astral" href="http://astral.berkeley.edu/index.html" 
target="_blank"><b>ASTRAL Compendium</b></a>
<dd>
Allows <a href="UsersGuide/fetch.html">fetching</a> 
domain structures specified by
<a href="UsersGuide/scop.html">SCOP domain identifier</a>.
<br><br>

<dt><a name="viperdb" 
href="http://viperdb.scripps.edu/" target="_blank"><b>Virus
Particle Explorer database</b></a>
<dd>
Allows <a href="UsersGuide/fetch.html">fetching</a> 
icosahedral virus capsid structures (transformed into a standard
orientation) specified by PDB ID.
<br><br>

<dt><a name="modbase" 
href="http://modbase.compbio.ucsf.edu/modbase-cgi/index.cgi"
target="_blank"><b>ModBase</b></a>
<dd>
Allows <a href="UsersGuide/fetch.html">fetching</a> 
<i>modeled</i> structures 
(<a href="UsersGuide/modbase.html">comparative models</a>)
specified by SwissProt, TrEMBL, GenPept or PIR accession code.
<br><br>

<dt><a name="pqs" href="http://pqs.ebi.ac.uk/pqs-doc.shtml" 
target="_blank"><b>Protein Quaternary Structure server</b></a>
<dd>
Allows <a href="UsersGuide/fetch.html">fetching</a> 
<i>predicted</i> biological units specified by PDB ID.
<br><br>

<dt><a name="pub3d"><b>Pub3D</b></a>
<dd>
Allows <a href="UsersGuide/fetch.html">fetching</a> 
<i>modeled</i> organic molecule structures specified by
<a href="http://pubchem.ncbi.nlm.nih.gov/"
target="_blank">PubChem</a> CID.  The <b>Pub3D</b></a> database 
contains structures for millions of PubChem entries 
and is provided by the <a href="http://www.chembiogrid.org/index.html"
target="_blank">Chemical Informatics and Cyberinfrastructure Collaboratory
(CICC)</a> at Indiana University. This database is described in 
<a href="http://www.biomedcentral.com/1471-2105/8/487"
target="_blank">Willighagen <i>et al.</i></a>, <i>BMC Bioinformatics</i>
<b>8</b>:487 (2007).
<br><br>

<dt><a name="smi23d" href="http://cheminfo.wikispaces.com/smi23d"
target="_blank"><b>smi23d</b></a> 
<dd>
Generates <i>modeled</i> organic molecule structures specified by
<a href="http://www.daylight.com/dayhtml/doc/theory/theory.smiles.html"
target="_blank">SMILES</a> string with the command
<a href="UsersGuide/midas/open.html"><b>open</b></a> or in
<a href="ContributedSoftware/editing/editing.html"><b>Build Structure</b></a>.
The <b>smi23d</b> service deploys the same procedure used to populate 
<b>Pub3D</b> (<a href="#pub3d">above</a>) and is provided by the
<a href="http://www.chembiogrid.org/index.html"
target="_blank">Chemical Informatics and Cyberinfrastructure Collaboratory
(CICC)</a> at Indiana University.
<dd>

</blockquote>

Density maps:
<blockquote>
<dt><a name="eds" href="http://eds.bmc.uu.se/eds/index.html" 
target="_blank"><b>Electron Density Server</b></a>
<dd>
Allows <a href="UsersGuide/fetch.html">fetching</a> 
electron density maps (2fo-fc and fo-fc) specified by PDB ID.
</blockquote>

Other data or calculations:
<blockquote>

<dt><a name="castp" href="http://sts-fw.bioengr.uic.edu/castp/index.php" 
target="_blank"><b>Computed Atlas of Surface Topography of proteins</b></a>
(CASTp)
<dd>
Allows <a href="UsersGuide/fetch.html">fetching</a> 
precomputed <a href="UsersGuide/castp.html">pocket measurements</a> 
and corresponding structures specified by PDB ID.
<br><br>

<dt><a name="structools" href="http://helixweb.nih.gov/structbio/basic.html" 
target="_blank"><b>StrucTools</b></a>
<dd>
Used by <a href="ContributedSoftware/surfvol/surfvol.html"><b>Area/Volume
from Web</b></a> to calculate atomic surface areas (various methods)
and Voronoi volumes from coordinates.
<br><br>

<dt><a name="blast"><b>Protein BLAST</b></a>
<dd>
The <a href="UsersGuide/blast.html"><b>Blast Protein</b></a> tool
performs protein sequence searches using a
<a href="http://blast.ncbi.nlm.nih.gov/" target="_blank">BLAST</a>
Web service hosted by the
<a href="http://www.cgl.ucsf.edu/Overview/index.html" target="_blank">UCSF
Resource for Biocomputing, Visualization, and Informatics (RBVI)</a>.
<br><br>

<dt><a name="gifcreator"><b>GIF Creator for Chemical Structures</b></a>
<dd>
Used by <a href="ContributedSoftware/diagram/diagram.html"><b>Structure
Diagram</b></a> and
<a href="ContributedSoftware/viewdock/framevd.html" 
target="_top"><b>ViewDock</b></a>
to draw 2D diagrams of small organic molecules given coordinates or a
<a href="http://www.daylight.com/dayhtml/doc/theory/theory.smiles.html"
target="_blank">SMILES</a> string.  The Web service is courtesy of 
<a href="http://www2.ccc.uni-erlangen.de/people/Wolf_Ihlenfeldt/"
target="_blank">Dr. W. D. Ihlenfeldt</a> 
(see <a href="ContributedSoftware/diagram/diagram.html#conditions">usage 
conditions</a>) and hosted by the 
<a href="http://www.cgl.ucsf.edu/Overview/index.html" target="_blank">UCSF
Resource for Biocomputing, Visualization, and Informatics (RBVI)</a>.

</blockquote>

<a name="webdatasites">
<h2>Sites That Provide Chimera Web Data</h2></a>

<p>
Accessing
<a href="ContributedSoftware/webdata/chimerax.html">Chimera Web data</a>
(such as by clicking a link or choosing a display option at one of these
sites) automatically displays database contents or server results
in a locally installed copy of Chimera.
</p>

<blockquote>

<dt><a name="genomebrowser" href="http://genome.ucsc.edu/"
target="_blank"><b>UCSC Genome Browser</b></a>
<dd>
Chimera links are provided in UCSC Gene pages and SNP track details pages.
Clicking a Chimera link displays the corresponding structure 
with nonsynonymous SNP residues (if any) shown with color and 
labeled with the dbSNP identifier.
<br><br>

<dt><a name="consurf" href="http://consurf.tau.ac.il/"
target="_blank"><b>The ConSurf Server</b></a>
<dd>
Clicking the link to view results in Chimera displays the query structure
and the multiple sequence alignment from ConSurf, both colored by the
server's conservation measure.  The ConSurf-calculated phylogenetic tree 
and custom headers are shown along with the sequence alignment.
<br><br>

<dt><a name="modbase2" 
href="http://modbase.compbio.ucsf.edu/modbase-cgi/index.cgi"
target="_blank"><b>ModBase</b></a>
<dd>
Choosing the "launch Chimera" option
to show a modeled structure displays it along with the
template structure and their pairwise sequence alignment.
<br><br>

<dt><a name="sfld"
href="http://sfld.rbvi.ucsf.edu" target="_blank"><b>The Structure-Function 
Linkage Database</b></a>
<dd>
Chimera Web data files are provided for viewing enzyme structures, 
active sites (Chimera sessions), and sequence alignments.

</blockquote>

</body>
</html>
