<!doctype html public "-//W3C//DTD HTML 4.0//EN">
<!-- vim:sw=1:
>>> UCSF Chimera Copyright
Copyright (c) 2009 Regents of the University of California.
All rights reserved.  This software provided pursuant to a
license agreement containing restrictions on its disclosure,
duplication and use.  This notice must be embedded in or
attached to all copies, including partial copies, of the
software or any revisions or derivations thereof.
<<< UCSF Chimera Copyright
-->
<html>
 <head>
  <meta http-equiv="Content-Type" content="text/html; charset=ISO-8859-1">
  <link rel="icon" href="../chimera16.png" type="image/png">
  <title>UCSF Chimera Version 1.4 Build 29530 Release Notes</title>
 </head>

 <body>
  <p>
  <img src="../titleChimera.png" width="370" height="55" alt="logo">
  </p>

<font size="-1">
<a href="http://www.cgl.ucsf.edu/chimera/docs/relnotes.html">Release History</a>
</font>

  <h2>UCSF Chimera Version 1.4.1 Build xxxx Release Notes (xx Jan 2010)</h2>

<hr>
  <h3>Changes since release 1.4:</h3>

   <h4>Major Bugs Fixed:</h4>
   <ul>
    <li>the problem in Chimera version 1.4 (Mac Aqua only) of a hang
    caused by trying to access the menu while running a background job 
    (e.g. Blast search) has been fixed.  Blast search options that
    were disabled in the Mac Aqua version have been re-enabled.
   </ul>
<hr>

  <h3>Changes in release 1.4 relative to 
   <a href="1.3.html">previous production release (1.3)</a>:</h3>

  <blockquote>

   <h4>Appearance and Image-Saving:</h4>
    <ul>
    <li> nicer initial default display
    (<b>New Molecules</b> "smart" preference, similar to ribbons preset; 
    can be turned off to give the previous behavior, or circumvented 
    with command <b>open noprefs</b>)
    <li> default ribbon style changed to rounded 
    (<b>New Molecules</b> preference) 
    <li> CA-only and P-only structures can be shown as ribbons and worms
    <li> ribbons preset includes special nucleotide representations,
    hides hydrogens on carbon atoms
    <li> default nucleic acid ribbon rotated 90&deg; from the previous default
    <li> improved transparent surface appearance (single-layer and 
     angle-dependence options, can be combined with silhouette edges)
    <li> quality option in <b>Lighting</b> tool turns on
    glossy lighting and angle-dependent transparency 
    (grayed out if your graphics driver is not capable)
    <li> BILD objects can be made transparent 
    (thanks to David Beck, University of Washington)
    <li> images now rendered offscreen on most platforms (not supported by 
    X11 in Mac OS or by certain older machines with other operating systems)
    <li> raytracing:
      <ul>
      <li>global clipping included
      <li>improved translation of transparent colors from Chimera to POV-Ray
      <li>defaults changed: increased antialias threshold (faster rendering),
        no jitter (better movie appearance)
      </ul>
    <li> added row-interleaved stereo mode in <b>Camera</b> tool
    for use with displays from VRex, Miracube, Zalman, and others
    </ul>

   <h4>I/O:</h4>
    <ul>
    <li> new <b>Fetch</b> preferences allow specifying a download directory, 
    later using local copies of the data (when available)
    <li> added fetch types <b>EMDB</b> and <b>EMDB & fit PDBs</b>
    <li><b>Fetch by ID</b> allows keyword search of EMDB
    <li> Pub3D fetching and SMILES&rarr;3D conversion updated to use the new
    REST services provided by the Chembiogrid (CICC) at Indiana University
    <li> output TIFF images now LZW-compressed 
    (uncompressed available as "TIFF-fast")
    <li> added output image format PPM
    <li> added export format STL (binary, triangle-based)
    <li> added <b>New Molecules</b> preference 
    to control naming of models read from Mol2 files
    <li> sessions <i>saved</i> with this release are much more memory
   efficient as they are being restored (but approximately the same
   memory use after restoration is complete)
    <li> session files containing surfaces are smaller
    <li> exported X3D files now default to version 3.0, unless version 3.2
    features are needed (<i>e.g.</i>, clip planes, orthographic viewpoints),
    so more X3D browsers and converters can understand the files
    </ul>

   <h4>General:</h4>
   <ul>
    <li> new <b>front center</b> rotation method is the default
    <li> clipping is off by default (front/back planes automatically adjust to
    avoid clipping anything) but can be turned on by moving the planes 
    or focusing on specific items
    <li> center of rotation rather than camera center used for 
    zooming with the mouse or the <b>scale</b> command
    <li> labels can be repositioned with the mouse (default Ctrl-right button)
    <li> amino acid <b>phi</b>, <b>psi</b>, <b>chi</b><sub>N</sub> angles
    automatically assigned as attributes and can be changed with
    <b>setattr</b> or in the <b>Selection Inspector</b>
    <li> new preferences: initial window size, how often to check for updates
    <li> Matplotlib included in distribution
    <li> on Macs, all file types can be opened by drag-and-drop on the 
    Chimera icon, not just PDB and Chimera Web data
    <li> on Mac Aqua, click-through has been implemented and the "window
    focus follows mouse" preference removed
    <li> can read double-precision DelPhi maps produced by some versions 
    of the Intel Mac executable from the Honig lab
    <li> 3Dconnexion SpaceNavigator<small><sup>TM</sup></small>
    input device supported on Linux (thanks to 
    Thomas Margraf, University of Hamburg)
   </ul>

   <h4>New Tools:</h4>
   <ul>

   <li><b>Add Ions</b> (Structure Editing, Amber) &mdash; 
    add counterions using AmberTools (thanks to Wei Zhang, 
    The University of Texas Health Science Center at Houston)

    <li><b>Blast Protein</b> (Sequence) &mdash;
    perform protein sequence searches using a BLAST Web service 
    hosted by the UCSF
    Resource for Biocomputing, Visualization, and Informatics (RBVI);
    the pseudoalignment from BLAST can be shown in <b>Multalign Viewer</b>, 
    and any corresponding structures can be fetched and superimposed

   <li><b>Coulombic Surface Coloring</b> (Surface/Binding Analysis) &mdash; 
    color molecular surfaces by Coulombic electrostatic potential
    (see <a href="http://tinyurl.com/mzopva" 
    target="_top">http://tinyurl.com/mzopva</a>
    for an informal comparison with published figures of
    Poisson-Boltzmann electrostatic potential)

   <li><b>Hide Dust</b> (Volume Data) &mdash; 
    hide small disconnected bits of a surface

   <li><b>Task Panel</b> (General Controls) &mdash; 
   interface to jobs started by Chimera

   <li><b>Thermal Ellipsoids</b> (Structure Analysis) &mdash; 
   show anisotropic B-factors
   </ul>

   <h4>Tool Changes:</h4>
   <ul>

    <li><b>2D Labels</b> and <b>Color Key</b> (Utilities) &mdash; 
      <ul>
      <li>added font style/typeface options
      <li>checkboxes allow showing/hiding individual 2D labels
      </ul>

    <li><b>Add Charge</b> (Structure Analysis, Structure Editing) &mdash; 
    charge methods for nonstandard residues shown on initial dialog

    <li><b>Axes/Planes</b> (formerly <b>Axes</b>) (Structure Analysis) &mdash; 
      <ul>
      <li>axis radii can be set to the average axis-atom distances
      <li>axis colors shown in color wells, can be changed
      <li>planes can be defined, displayed as disks, included in measurements
      </ul>

    <li><b>Build Structure</b> (Structure Editing) &mdash; secondary dialog 
    allows more convenient specification of peptide phi/psi angles

    <li><b>Dock Prep</b> (Structure Editing) &mdash; added options for
    changing modified RNA residues 5BU, CSL, UMS to the corresponding 
    standard residues

    <li><b>Effects</b> (Viewing Controls) &mdash; depth cueing 
    <b>yon intensity</b> parameter replaced with <b>end</b> parameter
    for greater control when clipping is off

    <li><b>Fit in Map</b> (Volume Data) &mdash; 
       <ul>
       <li>Undo/Redo capability added
       <li>map-map correlation can be optimized
       <li>movement can be restricted to selected atoms only (previously
         the entire models containing the atoms were moved as rigid bodies)
       </ul>
 
    <li><b>Icosahedron Surface</b> (Higher-Order Structure) &mdash; 
    status saved in sessions

    <li><b>Lens Inspector</b> (Depiction) removed

    <li><b>MatchMaker</b> (Structure Comparison) &mdash; 
     added option to call <b>Match -> Align</b> after superposition
     to create a structure-based alignment

    <li><b>Match -> Align</b> (Structure Comparison, Sequence) &mdash; 
      <ul>
      <li>RMSDs over fully populated columns of the sequence alignment
      reported in Reply Log
      <li>can iterate refitting the structures using the sequence alignment 
     and generating a new sequence alignment from the new superposition
      </ul>

    <li><b>MD Movie</b> (MD/Ensemble Analysis) &mdash; 
      <ul>
      <li>added support for Gromacs 4 trajectories 
      <li>Amber residue names remapped to standard PDB residue names for better
    integration with other tools (Rotamers, Nucleotides, Sequence, etc.)
      </ul>

    <li><b>Model Panel</b> (General Controls) &mdash; 
     <ul>
     <li>columns reordered, narrowed by shortening Active and Shown to A and S
     <li>arbitrary notes can be created and shown as a column
     <li>models can be renamed
     </ul>

    <li><b>Movement Mouse Mode</b> (Movement) &mdash; 
      added option to move entire molecule model under cursor

    <li><b>Multalign Viewer</b> (aka <b>Sequence</b>) &mdash; 
     <ul>
     <li>Preferences reorganized, includes separate Appearance settings 
     for multiple alignments and single sequences
     <li>allows copying a sequence as plain text so it can be pasted 
     into another application window
     <li><b>Sequence</b> includes any residues not in the structure itself
     but in SEQRES records in the input structure file
     <li>moved to new tools category <b>Sequence</b>
     </ul>

    <li><b>Nucleotides</b> (Depiction) &mdash
     <ul>
     <li>new ladder display and associated parameters
     <li>interface reorganized, defaults changed
     <li>representation colors automatically update to match atoms
     <li>representations saved in sessions
     </ul>

    <li><b>Reply Log</b> (Utilities) &mdash; 
     text-copying and string-searching added to dialog

    <li><b>Surface Color</b> (Volume Data) &mdash; 
     <ul>
     <li>coloring by volume data has per-pixel option
     <li>coloring by electrostatic potential uses surface offset 
     </ul>

    <li><b>Transform Molecule Coordinates</b> (Movement) &mdash; 
     added buttons to <b>Reset</b> (restore original coordinates) 
     and <b>Set</b> (equivalent to reset + apply)

    <li><b>Unit Cell</b> (Higher-Order Structure) &mdash; 
     can make blocks of multiple unit cells

    <li><b>Volume Filter</b> (Volume Data) &mdash; 
     <ul>
     <li>renamed from <b>Gaussian Filter</b>
     <li>additional options for binning, smoothing, shifting, scaling data
     </ul>

    <li><b>Volume Tracer</b> (Volume Data) &mdash; 
     <ul>
     <li>markers placed in successive residues for 
      easier command-line specification
     <li>added option to place markers on surfaces
     </ul>

    <li><b>Volume Viewer</b> (Volume Data) &mdash; 
     <ul>
     <li>outline box linewidth can be adjusted
     <li>initial thresholds for signed data are placed symmetrically about zero
     <li>region autodisplay setting moved from <b>Data display options</b> to
     <b>Subregion selection</b>
     </ul>

   </ul>

   <h4>New Commands:</h4>
   <ul>
    <li><b>aniso</b> &mdash; show thermal ellipsoids
    <li><b>aromatic</b> &mdash; show ring aromaticity
    <li><b>coordset</b> &mdash; play through frames of a trajectory
    <li><b>define</b> &mdash; calculate axes, planes for sets of atoms
    <li><b>fillring</b> &mdash; show rings as filled
    <li><b>fly</b> &mdash; smoothly traverse a series of saved positions
    <li><b>mcopy</b> &mdash; copy settings (colors, display styles, etc.)
    from one molecule model to another
    <li><b>measure</b> &mdash; perform several calculations on structures,
        surfaces, and maps (relative transformation, surface area, enclosed
        volume, map-map correlation, inertia ellipsoid, etc.)
    <li><b>nucleotides</b> &mdash; create special nucleotide representations
     (command implementation of <b>Nucleotides</b> tool)
    <li><b>pause</b> &mdash; suspend command script execution until the
        user presses a key
    <li><b>runscript</b> &mdash; run Python script with command-line arguments
    <li><b>scolor</b> &mdash; color surfaces a single color 
         or by volume data or geometry
         (command implementation of <b>Surface Color</b>)
    <li><b>sop</b> &mdash; edit a surface model
    <li><b>tile</b> &mdash; arrange models in a plane
    <li><b>windoworigin</b> &mdash; set graphics window location
   </ul>

   <h4>Command Changes:</h4>
   <ul>
    <li> command script execution can be aborted with <b>Esc</b>, 
    or paused/resumed with <b>Shift-Esc</b>
    <li> commas in command-line atom specifications imply ordering
    (previously, comma-separated models, residues, or atoms were not 
    necessarily used in the order in which they were typed)
    <li> more commands allow colors to be specified with Tk codes
    (e.g., #ff00ff) or R,G,B[,A] tuples 
    <li><b>2dlabels</b> &mdash; added font style/typeface options
    <li><b>addcharge</b> &mdash; default charge set changed to ff99SB
    <li><b>cofr</b> &mdash; fixed center of rotation
   can be specified in the coordinate system of a reference model; syntax
   improvements, access to all center of rotation methods
    <li><b>focus</b> &mdash; added <b>~focus</b> 
   (equivalent to <b>focus</b> without arguments)
    <li><b>label</b>, <b>rlabel</b> &mdash; label offset can be specified
    <li><b>mask</b> &mdash; can invert mask to get the opposite data
    <li><b>match</b> &mdash; added options to suppress the transformation, 
    report the matrix
    <li><b>matrixcopy</b> &mdash; multiple target models can be specified
    <li><b>move</b> &mdash; translation direction can be specified in
   the coordinate system of a reference model; specified models move
   regardless of whether they are active
    <li><b>rangecolor</b> &mdash; added option to call <b>Color Key</b>
    <li><b>reset</b> &mdash; models not present when the position was saved
   now transform along with the lowest-ID model that was present
    <li><b>rock, roll, turn</b> &mdash; rotation axis and center
   can be specified in the coordinate system of a reference model;
   specified models move regardless of whether they are active
    <li><b>rotation</b> &mdash; 
    can specify torsion ID, adjust torsions (rotate bonds), deactivate torsions
    <li><b>set</b> &mdash; expanded to control depth-cueing, silhouette edges,
    subdivision quality, projection mode, lighting quality
    <li><b>setattr</b> &mdash; expanded to control molecular surface attributes
    <li><b>shape</b> &mdash; 
      <ul>
      <li>location (center) can be specified with x,y,z coordinates 
      or a set of atoms
      <li>slab option added
      <li>cylinders can be capped
      <li>can make a tube connecting a series of atoms or markers 
      (&agrave; la <b>Volume Tracer</b>)
      </ul>
    <li><b>surftransparency</b> &mdash; expanded to nonmolecular surfaces;
    frame argument added for movie scripting
    <li><b>sym</b> &mdash; cyclic, dihedral, icosahedral, helical, and
    translational symmetries can be specified
    (previously only BIOMT records were used)
    <li><b>volume</b> &mdash; 
       <ul>
       <li>added planes playback for scripting animations
       <li>file-saving allows step, subregion specification, 
       saving multiple files at once
       </ul> 
    <li><b>vop</b> &mdash;
       <ul>
       <li>added several types of filtering, binning, scaling,
     axis permutation, subregion (box) extraction, periodic map extension, 
     map addition, morphing
      <li>subtraction improved to automatically resample as needed,
     use specified scale factors
      </ul>
   <li>redundant commands deprecated: <b>load</b>, <b>source</b>,
   <b>x3dsave</b>, <b>brotation</b>, <b>push</b>, <b>pop</b>
   </ul>

   <h4>New Tutorials:</h4>
   <ul>
    <li><b>Density Display</b> image tutorial
    <li><b>Similar Binding Sites</b> image tutorial
    <li><b>B-Factor Coloring</b> image tutorial
   </ul>

   <h4>Major Bugs Fixed:</h4>
   <ul>
    <li>interpolation between saved positions by <b>reset</b>
    corrected to better preserve the relative orientations of models
    as they move
    <li>the color menu in the Mac Aqua version now shows colors 
   (not just their names) and has two columns
   <li>Image annotations (Author, Copyright, <i>etc</i>.) for PNG files
   are now saved correctly &mdash; only JPEG files are missing the annotations.
   </ul>

   <h4>Items Scheduled for Removal:</h4>
   <ul>
   <li>Some features previously scheduled for removal after release 1.3 
   are still included, but are slated for removal before the 1.5 release: 
   <b>Minrms Plot</b> (to be distributed with MinRMS instead),
   <b>Selector Construction Panel</b>,
   the low-resolution display option in <b>Side View</b>,
   coloring capabilities in <b>ResProp</b>,
   and the "named regions" and "precomputed subsamples" parts of the
   <b>Volume Viewer</b> dialog
   <li><b>Phantom Force Feedback</b> is also scheduled for removal before the
   1.5 release
   </ul>
    

   <h4>Known Bugs:</h4>
   <ul>
    <li>in Mac Aqua Chimera, trying to access the menu while a background
    job (e.g. Blast search) is running causes a hang.
    <b>A 1.4.1 release that fixes this problem is planned.</b>
    In the meanwhile, multiple-iteration blast searches (psiblast)
    and blasting against nr are disabled in the Mac Aqua version only.
    <li>Mac OS 10.5 / X11 problems:
       <ul>
       <li> sometimes display does not update
       (can be forced by slightly moving the view contents with the mouse)
       <li> sometimes saved images have a corrupt or "mis-tiled" appearance.
        Workarounds to try include:
         <ul>
         <li>showing the <b>Command Line</b> if it is not shown 
        (or if it is shown, hiding it) and then re-saving the image
         <li>saving the position with command <b>savepos</b>, jiggling
         the view with the mouse, restoring the position with <b>reset</b>,
         and then re-saving the image
         <li>saving a session, restoring the session in the Mac Aqua
         version of Chimera, and then re-saving the image
         </ul>
        </ul>
     <li>surfaces fail for some structures
     <li>POV-Ray incorrectly renders some of the characters in the fonts
     Chimera uses, especially the zero
   </ul>

  </blockquote>
 </body>
</html>
