<!--
--- UCSF Chimera Copyright ---
Copyright (c) 2003-2009 Regents of the University of California.
All rights reserved.  This software provided pursuant to a
license agreement containing restrictions on its disclosure,
duplication and use.  This notice must be embedded in or
attached to all copies, including partial copies, of the
software or any revisions or derivations thereof.
--- UCSF Chimera Copyright ---
-->

<html>
<head>
<TITLE>MatchMaker</TITLE>
<style type="text/css">
A.nounder:link {text-decoration: none}
A.nounder:visited {text-decoration: none}
A.nounder:active {text-decoration: none}
A.nounder:hover {text-decoration: underline; color: red;}
</style>
</head>
<body>
<h3>MatchMaker <img src="mmicon.png" alt="MatchMaker icon"></h3>
<p>
<b>MatchMaker</b> superimposes protein or nucleic acid structures
by first creating pairwise sequence alignments, 
then fitting the aligned residue pairs.
Residue types and/or secondary structure information 
can be used to create the initial sequence alignments.
Fitting uses <a href="#matchatom">one point</a> per residue.
Optionally, a structure-based multiple sequence alignment can be computed
after the structures have been superimposed.
</p><p>
Note: if it is already known which residue numbers in one structure 
should be paired with which residue numbers in the other, 
another possibility is to use the command 
<a href="../../UsersGuide/midas/match.html"><b>match</b></a>.
See <a href="../../UsersGuide/superposition.html">superimposing
structures</a> for a discussion of the different methods available in Chimera.
See also:
<a href="../matchalign/matchalign.html"><b>Match -> Align</b></a>,
<a href="../multalignviewer/framemav.html" target="_top"><b>Multalign
Viewer</b></a>,
the <a href="../../UsersGuide/tutorials/alignments.html">Superpositions and
Alignments tutorial</a>, and
<blockquote>
<a href="http://www.biomedcentral.com/1471-2105/7/339" target="_blank">
Tools for integrated sequence-structure analysis with UCSF Chimera.</a>
Meng EC, Pettersen EF, Couch GS, Huang CC, Ferrin TE.
<i>BMC Bioinformatics.</i> 2006 Jul 12;7:339.
</blockquote>
</p><p>
There are <a href="../../UsersGuide/extension.html">several ways to start</a>
<b>MatchMaker</b>, a tool in the <b>Structure Comparison</b> category.
<b>MatchMaker</b> is also implemented as the command
<a href="../../UsersGuide/midas/mmaker.html"><b>mmaker</b></a>
(or <a href="../../UsersGuide/midas/matchmaker.html"><b>matchmaker</b></a>).
</p><p>
<a name="top">
The <b>MatchMaker</b> dialog is organized by the main steps to be performed:
</a>
<ol>
<li><a href="#initial">generating pairwise sequence alignments</a>
<li><a href="#matching">matching</a>, <i>i.e.</i>,
superimposing the structures according to those pairwise alignments
<li>optionally, <a href="#final">creating a multiple sequence alignment</a>
from the structural superposition
</ol>
<b>Save settings</b> writes the current <b>MatchMaker</b> parameters to the
<a href="../../UsersGuide/preferences.html#preffile"><b>preferences</b> 
file</a>. <b>Reset to defaults</b> resets the dialog to the factory default 
parameter settings without changing any preferences.
</p><p>
Clicking <b>OK</b> or <b>Apply</b> will start the calculations
with or without closing the dialog, respectively.
Sequence alignment scores, parameter values, and structure RMSDs will be 
reported in the <a href="../../UsersGuide/reply.html"><b>Reply Log</b></a>.
</p><p>
<b>Cancel</b> simply closes the dialog, while <b>Help</b> opens
this manual page in a browser window.
</p>

<h3><a href="#top" class="nounder">&larr;</a>
<a name="initial">Initial Pairwise Sequence Alignments</a></h3>
<p>
<a name="pairing"><b>Chain pairing</b></a> options:
<ul>
<li><b>Best-aligning pair of chains between reference and match structure</b> 
(default)
- One reference structure and one or more structures to match
should be chosen.  For each structure to be matched, the
reference-match pair of chains with the highest 
<a href="#scoring">sequence alignment score</a> will be used.
<li><b>Specific chain in reference structure
with best-aligning chain in match structure</b>
- One reference chain and one or more structures to match should be chosen.  
Individual lines or blocks of lines can be chosen with the left 
mouse button; Ctrl-click toggles the status of a line.
For each structure to be matched,
the chain that aligns to the reference chain with the highest
<a href="#scoring">sequence alignment score</a> will be used.
<li><b>Specific chain(s) in reference structure
with specific chain(s) in match structure</b>
- One or more reference chains should be chosen from the list.
For each reference chain chosen, one chain to be matched should 
be chosen from the corresponding pulldown menu.  If multiple chains are to
be matched to the same reference chain, it is necessary to match them
in separate steps (by choosing the chain to match and then clicking
<b>Apply</b>).  A given chain cannot be matched to two different
reference chains simultaneously, and chains from the same structure 
(molecule model)
cannot simultaneously serve as a reference chain and a chain to match.
</ul>
<a name="algorithm"><b>Alignment algorithm</b></a>:
  <ul>
  <li><b>Needleman-Wunsch</b> (default) - global
  <li><b>Smith-Waterman</b> - local
  </ul>
<a name="scoring">
Sequence alignment scoring can include a residue similarity term,
a secondary structure term, and gap penalties.</a>
<ul>
<li><b>Matrix</b> (default <b>BLOSUM-62</b>) - what
<a href="../../UsersGuide/midas/mmaker.html#matrix">substitution matrix</a>
to use for the residue similarity part of the score.
If an amino acid matrix is chosen, only peptide sequences
will be aligned; if a nucleic acid matrix is chosen, only
nucleic acid sequences will be aligned.  An error message will appear
if there are no reference-match pairs of the appropriate type.
<br><br>
<li>Gap penalties: When <a href="#useSS">secondary structure scoring</a>
is included, the secondary-structure-specific <b>Gap opening penalties</b> 
(<b>Intra-helix</b>, <b>Intra-strand</b>, <b>Any other</b>)
are used instead of the single <b>Gap opening penalty</b>.
The same <b>Gap extension penalty</b> is used, however.
<br><br>
<li><a name="useSS"><b>Include secondary structure score (<i>N</i>%)</b></a>
(default <b>on</b> and <b><i>N</i></b>=<b>30</b>)
- whether to include a secondary structure term in the score, 
and at what weight relative to the residue similarity term.
<b>Show parameters</b> reveals the secondary structure scoring parameters.
<b><i>N</i></b> reflects the relative weights of the terms,
which can be adjusted by moving the slider.  
If the weight is 30%, for example,
<blockquote>
total score = 0.70(residue similarity score) + 0.30(secondary structure score)
&ndash; gap penalties
</blockquote>
Setting the weight to 0% is <a href="#notes">not the same</a>
as turning the option off, however.
The values in the secondary structure <b>Scoring matrix</b> 
(for all pairwise combinations of <b>H</b> helix, <b>S</b> strand, and 
<b>O</b> other) and the secondary-structure-specific 
<b>Gap opening penalties</b> can be adjusted.
<b>Reset secondary structure scoring parameters to defaults</b>
can be used to restore the default values of all secondary structure
scoring parameters.  
<br><br>
<li><a name="computeSS"><b>Compute secondary structure assignments</b></a>
(available when 
<a href="#useSS">secondary structure scoring</a> is used; default <b>on</b>)
- whether to first identify helices and strands by running the
<a href="../../UsersGuide/midas/ksdssp.html"><b>ksdssp</b></a> algorithm,
overwriting any pre-existing secondary structure assignments.
A reason to use this option despite existing assignments
is that the use of consistent criteria tends to improve the results.
Pre-existing secondary structure assignments may have been determined 
using different methods or different parameters for different structures.
<a href="../../UsersGuide/midas/ksdssp.html"><b>Ksdssp</b></a> 
parameter defaults can be adjusted with the
<a href="../../UsersGuide/modelpanel.html#computess"><b>compute SS</b></a>
dialog (opened from the 
<a href="../../UsersGuide/modelpanel.html"><b>Model Panel</b></a>).
<br><br>
<li><b>Show pairwise alignment(s)</b> (default <b>off</b>)
- whether to display the resulting pairwise 
reference-match sequence alignments; each will be shown in a separate 
<a href="../multalignviewer/framemav.html" target="_top">
<b>Multalign Viewer</b></a> window.
When <a href="#iterate">fit iteration</a> is employed,
the pairs used in the final fit will be shown in the alignment as a
<a href="../multalignviewer/multalignviewer.html#regions">region</a>
(colored boxes) named <b>matched residues</b>.  The
<a href="../multalignviewer/multalignviewer.html#headers">header</a>
named <b>RMSD</b> is automatically shown and other headers hidden.  
This line shows the spatial variation among residues associated with
a column.  In the pairwise case, the value is simply the distance between
<a href="#matchatom">atoms</a> in the two residues associated with a column.
</p><p>
*These pairwise sequence alignments can be considered a by-product 
of superposition.
Successful superposition only requires these alignments to be partly
correct, as incorrect portions tend to be omitted during 
<a href="#iterate">fit iteration</a>.  
If the sequences are easy to align (highly similar),
the sequence alignments are likely to be correct throughout.
However, if the sequences are more distantly related, 
parts of the alignments may be incorrect even when a successful
superposition is produced.  In those cases, a
<a href="#final">structure-based alignment</a> should be superior.
</ul>
</p>

<h3><a href="#top" class="nounder">&larr;</a>
<a name="matching">Matching</a></h3>
<p>
<a name="matchatom">
Fitting uses one point per residue: CA atoms in amino acids and
C4' atoms in nucleic acids.</a>  
If a nucleic acid residue lacks a C4' atom
(some lower-resolution structures are P traces),
its P atom will be paired with the P atom of the aligned residue.
<blockquote>
<a name="iterate">
<b>Iterate by pruning long atom pairs until no pair exceeds [<i>x</i>]
angstroms</b></a>
(default <b>on</b> and <b><i>x</i></b>=<b>2.0</b>)  
- whether to iteratively remove far-apart residue pairs from
the "match list" used to superimpose the structures.  This does not
change the initial sequence alignment, but restricts which columns of
that alignment will be used in the final fit.
Otherwise, all of the columns containing both sequences
(i.e. without a gap) will be used.  In each cycle of iteration,
atom pairs are removed from the match list and the remaining
pairs are fitted, until no matched pair is
more than <b><i>x</i></b> &Aring; apart.
The atom pairs removed are either the 10% farthest apart of all pairs
or the 50% farthest apart of all pairs exceeding the cutoff, whichever
is the lesser number of pairs.  
Iteration tends to exclude sequence-aligned but conformationally dissimilar 
regions such as flexible loops, allowing a tighter fit of the 
best-matching "core" regions.
</blockquote>
Regardless of which chain(s) in a model to be matched are
aligned in sequence to the reference, the entire model will be reoriented.
</p>

<h3><a href="#top" class="nounder">&larr;</a>
<a name="final">Final Structure-Based Sequence Alignment</a></h3>
<p>
If <b>MatchMaker</b> is used simply to superimpose structures,
this step can be omitted.  
However, if one also wants a corresponding structure-based
sequence alignment, this step is recommended,
especially if the sequences are dissimilar.
<blockquote>
<b>After superposition, compute structure-based 
multiple sequence alignment</b> (default <b>off</b>)
- call <a href="../matchalign/matchalign.html"><b>Match -> Align</b></a> 
to generate a sequence alignment consistent with the superposition.  
If not called with this option,
<a href="../matchalign/matchalign.html"><b>Match -> Align</b></a> 
can still be started later independently.
</blockquote>
Calculating a structure-based alignment can take several minutes,
depending on the number of structures, but there are advantages:
<ul>
<li>it can provide better statistics for describing structural similarity
(RMSD, <i>etc.</i>) because more alignment columns are correct 
<li>it can produce a multiple sequence alignment, whereas
the initial sequence alignments are only pairwise
</ul>
The output sequence alignment is automatically shown in
<a href="../multalignviewer/framemav.html" target="_top">
<b>Multalign Viewer</b></a> and can be
<a href="../multalignviewer/multalignviewer.html#saving">saved to a file</a>
from that tool.	 The fully populated columns are highlighted as a
<a href="../multalignviewer/multalignviewer.html#regions">region</a>
(colored boxes).  Clicking the region will select the corresponding parts
of the structures, in effect their common cores.
The <a href="../multalignviewer/multalignviewer.html#headers">header</a>
named <b>RMSD</b> shows the spatial variation per column.
</p>

<a name="notes"><h3>Notes</h3></a>
<p>
<b>Meaning of 0% secondary structure score</b>.
Turning off <a href="#useSS"><b>Include secondary structure score</b></a>
is not the same as moving the slider to zero with the option turned on.
When the option is on:
<ul>
<li>The secondary-structure-specific gap opening penalties are used
regardless of the slider position.
<li>If <a href="#computeSS"><b>Compute secondary structure assignments</b></a>
is also turned on, 
<a href="../../UsersGuide/midas/ksdssp.html"><b>ksdssp</b></a> is run and
may change pre-existing secondary structure assignments.
</ul>
</p>

<hr>
<address>UCSF Computer Graphics Laboratory / July 2009</address>
</body>
</html>
