<!--
--- UCSF Chimera Copyright ---
Copyright (c) 2009 Regents of the University of California.
All rights reserved.  This software provided pursuant to a
license agreement containing restrictions on its disclosure,
duplication and use.  This notice must be embedded in or
attached to all copies, including partial copies, of the
software or any revisions or derivations thereof.
--- UCSF Chimera Copyright ---
-->

<html><head>
<title>Blast Protein</title>
</head><body>
<h3>Blast Protein <img src="blasticon.png" alt="Blast Protein icon"></h3>
<p>
<b>Blast Protein</b> performs protein sequence searches using 
a <a href="#notes">BLAST</a> Web service hosted by the
<a href="http://www.cgl.ucsf.edu/Overview/index.html" target="_blank">UCSF
Resource for Biocomputing, Visualization, and Informatics (RBVI)</a>.
Corresponding structures can be retrieved and automatically superimposed,
and the <a href="#notes">pseudo-multiple alignment</a> 
from BLAST can be shown in
<a href="../ContributedSoftware/multalignviewer/framemav.html"
target="_top"><b>Multalign Viewer</b></a>.
The search results are included in saved <a href="sessions.html">sessions</a>.
</p><p>
There are <a href="../../UsersGuide/extension.html">several ways to start</a>
<b>Blast Protein</b>, a tool in the <b>Sequence</b> category.
When it is started from the 
<a href="../ContributedSoftware/multalignviewer/multalignviewer.html#mavmenu-tools"><b>Tools</b> menu</a> in
<a href="../ContributedSoftware/multalignviewer/framemav.html"
target="_top"><b>Multalign Viewer</b></a>,
any of the sequences in the alignment can be specified as the query.
Otherwise, the query sequence can be:
<ul>
<li> <b>From Structure</b>
- a protein chain from a structure open in Chimera
<li> <a name="addplain"><b>Plain Text</b></a> 
- a protein sequence entered or pasted in the <b>Sequence</b> field 
as plain text
</ul>
Search parameters:
<ul>
<li><b>Program</b>
  <ul>
  <li><b>blast</b> (default) - protein-protein search 
  <li><b>psiblast</b> - iterated protein-protein search, where each pass 
  after the first uses a PSSM (position-specific scoring matrix) 
  based on the results of the previous pass
  </ul>
<li><b>Database</b>
  <ul>
  <li><b>pdb</b> (default) - sequences of structures in the
  <a href="http://www.rcsb.org" target="_blank">Protein Data Bank (PDB)</a>.
  Whereas completely identical sequences are collapsed to a single entry
  in the <b>pdb</b> sequence database from NCBI, the
<a href="http://www.cgl.ucsf.edu/Overview/index.html" target="_blank">RBVI</a>
  Web service uses a customized version of this database in which the
  identical sequences are separate entries.  The <b>List only...</b> option
  controls whether multiple hits with the same PDB identifier or only 
  the best-matching one should be included in the results.
  <li><b>nr</b> - all non-redundant GenBank CDS translations + 
  RefSeq Proteins + PDB + SwissProt + PIR + PRF.  This database is much
  larger than <b>pdb</b> and will take longer to search.
  </ul>
<li><b>E-value (1e-X)</b> - significance cutoff; 
  only matches with E-values &le; 10<sup>-X</sup> will be returned 
  (X=<b>3</b> by default) 
<li><b>Matrix</b> (BLOSUM45/<b>BLOSUM62</b>/BLOSUM80/PAM30/PAM70) 
  - amino acid substitution matrix to use for alignment scoring
<li><b>Passes</b> (default <b>1</b>) 
  - number of <b>psiblast</b> iterations; 1 pass is equivalent to <b>blast</b>,
  whereas multiple passes will find more distantly related sequences
<li><b>List only best-matching chain per PDB entry</b> (default)
- whether multiple hits with the same PDB identifier 
  (matches to different chains in that PDB entry) or only 
  the best-matching one should be included in the results
</ul>
Clicking <b>OK</b> launches the search and dismisses the dialog.
<b>Apply</b> launches the search without dismissing the dialog,
<b>Cancel</b> simply dismisses the dialog, 
and <b>Help</b> opens this manual page in a browser window.
</p><p>
The search is a background task.  
Clicking the information icon <img src="taskicon.png"> in the Chimera
<a href="chimerawindow.html#statusline">status line</a>
will bring up the
<a href="../ContributedSoftware/idle/task.html"><b>Task Panel</b></a>,
in which the job can be canceled if desired.
</p><p>
<a name="results">
Results are returned as a table of sequences 
with the following columns</a>:
<ul>
<li><b>GI</b> - sequence GI number
  (for the query, <b>model name</b> is shown instead) 
<li><b>PDB</b> - corresponding PDB identifier, if any
<li><b>Evalue</b> - significance value
<li><b>Score</b> - alignment score
<li><b>Description</b> - text description 
</ul>
One or more of the hits (lines in the table) can be chosen 
with the left mouse button.
A block of lines can be chosen by dragging, or by
clicking on the first (or last) line in the desired block
and then <b>Shift</b>-clicking on its last (or first) line.
<b>Ctrl</b>-click toggles the state (chosen or not) of single line.
<b>Show in MAV</b> displays the 
<a href="#notes">pseudo-multiple alignment</a> 
of the chosen sequences in 
<a href="../ContributedSoftware/multalignviewer/framemav.html"
target="_top"><b>Multalign Viewer</b></a>, and 
<b>Load Structure</b> retrieves and opens the PDB structures corresponding
to the chosen sequences (as described for
<a href="fetch.html"><b>Fetch by ID</b></a>)
and automatically superimposes them according to the
<a href="#notes">pseudo-multiple alignment</a>.
Entire PDB entries are retrieved, not just the matching chains.
</p><p>
<b>Quit</b> removes the table of results, whereas <b>Hide</b> undisplays
it without removing the data.  The hidden table can be shown again using the 
<b>Raise</b> option of its <a href="../../UsersGuide/instances.html">instance 
in the <b>Tools</b> menu</a>.
This is also useful when the table has become obscured by other windows.
</p>

<a name="notes">
<h3>Notes</h3></a>
<p>
<b>Basic Local Alignment Search Tool (BLAST)</b>.
The <a href="http://blast.ncbi.nlm.nih.gov/" target="_blank">BLAST</a>
software is provided by the 
<a href="http://www.ncbi.nlm.nih.gov/" target="_blank">NCBI</a>
and described in the following:
<blockquote>
<a href="http://nar.oxfordjournals.org/cgi/content/full/25/17/3389" 
target="_blank">Gapped BLAST and PSI-BLAST: 
a new generation of protein database search programs.</a>
Altschul SF, Madden TL, Sch&auml;ffer AA, Zhang J, Zhang Z, Miller W, Lipman DJ.
<i>Nucleic Acids Res.</i> 1997 Sep 1;25(17):3389-402.
<br><br>
<a href="http://www.ncbi.nlm.nih.gov/pubmed/2231712" target="_blank">
Basic local alignment search tool.</a>
Altschul SF, Gish W, Miller W, Myers EW, Lipman DJ.
<i>J Mol Biol.</i> 1990 Oct 5;215(3):403-10.
</blockquote>
</p><p>
<b>Pseudo-multiple alignment</b>.
The pseudo-multiple alignment from BLAST is not a true multiple alignment,
but a consolidation of the pairwise alignments of individual hits to the
query.  This output corresponds to the BLAST formatting option
(<a href="http://www.ncbi.nlm.nih.gov/BLAST/blastcgihelp.shtml#alignment_view"
target="_blank">alignment view</a>)
"flat query-anchored with letters for identities."
</p>
<hr>
<address>UCSF Computer Graphics Laboratory / November 2009</address>
</body></html>
