<!--
--- UCSF Chimera Copyright ---
Copyright (c) 2000-2009 Regents of the University of California.
All rights reserved.  This software provided pursuant to a
license agreement containing restrictions on its disclosure,
duplication and use.  This notice must be embedded in or
attached to all copies, including partial copies, of the
software or any revisions or derivations thereof.
--- UCSF Chimera Copyright ---
-->

<html>
<head>
<TITLE>Chimera Tools Index</TITLE>
</head>
<body>

<h2>Tools Index</h2>
<ol>
<li><a href="../ContributedSoftware/2dlabels/2dlabels.html">
<b>2D Labels</b></a>
- place text and
<a href="../ContributedSoftware/2dlabels/symbols.html">symbols</a>
in the display
<li><a href="../ContributedSoftware/addcharge/addcharge.html"
target="Content"><b>Add Charge</b></a>
- assign partial charges to atoms
<li><a href="../ContributedSoftware/addh/addh.html"><b>AddH</b></a>
- add hydrogens
<li><a href="../ContributedSoftware/addions/addions.html"><b>Add Ions</b></a>
- add monatomic counterions using
<a href="http://ambermd.org/#AmberTools" target="_blank">AmberTools</a>
<li><a href="../ContributedSoftware/structuremeas/structuremeas.html#adjust"
target="Content"><b>Adjust Torsions</b></a>
- rotate bonds (change dihedral angles)
<li><a href="../ContributedSoftware/structuremeas/structuremeas.html#angles"
target="Content"><b>Angles/Torsions</b></a>
- measure bond angles and dihedral angles
<li><a href="../ContributedSoftware/surfvol/surfvol.html" target="Content">
<b>Area/Volume from Web</b></a>
- use the <a href="http://helixweb.nih.gov/structbio/basic.html" 
target="_blank"><b>StrucTools</b> server</a> 
to calculate surface areas and Voronoi volumes
<li><a href="../ContributedSoftware/calculator/calculator.html"
target="Content"><b>Attribute Calculator</b></a>
- generate new numerical
<a href="../ContributedSoftware/defineattrib/defineattrib.html#attribdef">
attributes</a> from existing ones; calculate totals or averages of a given
numerical attribute
<li><a href="../ContributedSoftware/structuremeas/structuremeas.html#axes"
target="Content"><b>Axes/Planes</b></a>
- define geometric objects (axes, planes) based on sets of atoms,
perform related measurements
<li><a href="../ContributedSoftware/benchmark/benchmark.html"
target="Content"><b>Benchmark</b></a>
- measure hardware performance on standard Chimera rendering tasks
<li><a href="blast.html" target="Content"><b>Blast Protein</b></a>
- perform protein BLAST searches using a web service hosted by the 
<a href="http://www.cgl.ucsf.edu/Overview/index.html" target="_blank">UCSF
RBVI</a>
<li><a href="../ContributedSoftware/webdata/webdata.html#browser">
<b>Browser Configuration</b></a>
- configure web browsers to send Chimera
<a href="../ContributedSoftware/webdata/webdata.html#webdata">data
linked to web pages</a>
<li><a href="../ContributedSoftware/editing/editing.html"
target="Content"><b>Build Structure</b></a>
- generate atomic structures "from scratch" or modify existing molecules
<li><a href="sideview.html#camera" target="Content"><b>Camera</b></a>
- control viewing parameters
<li><a href="colortool.html" target="Content"><b>Color Editor</b></a>
- create colors interactively
<li><a href="../ContributedSoftware/2dlabels/2dlabels.html#colorkey">
<b>Color Key</b></a>
- create a color key for figures
<li><a href="../ContributedSoftware/colorss/colorss.html" target="Content">
<b>Color Secondary Structure</b></a>
- color peptides/proteins by secondary structure
<li><a href="../ContributedSoftware/colorzone/colorzone.html" 
target="Content"><b>Color Zone</b></a>
- color surfaces to match <a href="selection.html">selected</a> atoms,
split volume data by the resulting color zones
<li><a href="../UsersGuide/chimerawindow.html#emulator"><b>Command Line</b></a>
- enter Chimera <a href="framecommand.html" target="_top">commands</a>
<li><a href="../ContributedSoftware/constrained/constrained.html">
<b>Constrained Move</b></a>
- perform rotations and translations about specified axes
<li><a href="../ContributedSoftware/coulombic/coulombic.html"><b>Coulombic
Surface Coloring</b></a>
- color <a href="representation.html#surfaces">molecular surfaces</a>
by Coulombic electrostatic potential
<li><a href="../ContributedSoftware/crystalcontacts/crystalcontacts.html">
<b>Crystal Contacts</b></a>
- identify clashes between PDB symmetry copies
<li><a href="../ContributedSoftware/defineattrib/defineattrib.html"
target="Content"><b>Define Attribute</b></a>
- assign 
<a href="../ContributedSoftware/defineattrib/defineattrib.html#attribdef">
attribute</a> values to atoms, residues, or models
<li><a href="../ContributedSoftware/delphicontroller/delphicontroller.html">
<b>DelPhiController</b></a>
- interface to
<a href="http://wiki.c2b2.columbia.edu/honiglab_public/index.php/Software:DelPhi"
target="_blank">DelPhi</a> (obtained separately)
for calculating electrostatic potential
<li><a href="../ContributedSoftware/demos/demos.html"><b>Demos</b></a> and
<a href="../ContributedSoftware/demos/demos.html#editor"><b>Demo Editor</b></a>
- create and replay demonstrations in Chimera
<li><a href="../ContributedSoftware/structuremeas/structuremeas.html#distances"
target="Content"><b>Distances</b></a>
- measure distances between pairs of atoms
<li><a href="../ContributedSoftware/dockprep/dockprep.html"
target="Content"><b>Dock Prep</b></a>
- add hydrogens, charges, <i>etc.</i> to prepare structures for
<a href="http://dock.compbio.ucsf.edu/index.html" target="_blank">DOCK</a>
or for other calculations
<li><a href="sideview.html#effects" target="Content"><b>Effects</b></a>
- control visual effects such as depth cueing and silhouette edges
<li><a href="../ContributedSoftware/surfcolor/surfcolor.html"
target="Content"><b>Electrostatic Surface Coloring</b></a>
- color surfaces using an
<a href="filetypes.html#esp">electrostatic potential map</a>
(from any of several <a href="filetypes.html#esp">separate programs</a>)
<li><a href="../ContributedSoftware/ensemblecluster/ensemblecluster.html">
<b>Ensemble Cluster</b></a>
- cluster members of a conformational ensemble
<li><a href="../ContributedSoftware/ensemblematch/ensemblematch.html">
<b>Ensemble Match</b></a>
- match conformations from two ensembles
<li><a href="../ContributedSoftware/findclash/findclash.html"><b>Find 
Clashes/Contacts</b></a>
- identify clashes and/or contacts
<li><a href="../ContributedSoftware/findhbond/findhbond.html">
<b>FindHBond</b></a>
- find possible hydrogen bonds
<li><a href="../ContributedSoftware/fitmaps/fitmaps.html"><b>Fit in Map</b></a>
- fit atoms into a map (volume data) or one map into another
<li><a href="../ContributedSoftware/flatten/flatten.html"><b>Flatten 
Icosahedron</b></a>
- rearrange the faces of an icosahedral virus capsid into a plane
to create a paper model
<li><a href="../ContributedSoftware/volumeviewer/hidedust.html"><b>Hide
Dust</b></a>
- hide small disconnected bits of a surface
<li><a href="../ContributedSoftware/icosahedron/icosahedron.html">
<b>Icosahedron Surface</b></a>
- create a hybrid icosahedron/sphere surface 
<li><a href="../ContributedSoftware/idle/idle.html"><b>IDLE</b></a>
(Python <b>I</b>nteractive <b>D</b>eve<b>L</b>opment <b>E</b>nvironment)
- enter Python commands
<li><a href="../ContributedSoftware/intersurf/intersurf.html"
target="Content"><b>Intersurf</b></a>
- generate and display interface surfaces
<li><a href="../ContributedSoftware/accelerators/accelerators.html"
target="Content"><b>Keyboard Shortcuts</b></a>
- define and use keyboard shortcuts for Chimera functions
<li><a href="sideview.html#lighting" target="Content"><b>Lighting</b></a>
- adjust and save lighting parameters
<li><a href="../ContributedSoftware/matchalign/matchalign.html">
<b>Match -> Align</b></a>
- generate a sequence alignment from a structural alignment
<li><a href="../ContributedSoftware/matchmaker/matchmaker.html">
<b>MatchMaker</b></a>
- superimpose structures by first constructing a sequence alignment 
(optionally including secondary structure scoring) and then 
fitting the aligned residue pairs
<li><a href="../ContributedSoftware/movie/framemovie.html" target="_top">
<b>MD Movie</b></a>
- replay and analyze molecular dynamics trajectories
<li><a href="../ContributedSoftware/pickblobs/pickblobs.html">
<b>Measure and Color Blobs</b></a>
- color and measure disconnected parts of a surface
<li><a href="../ContributedSoftware/measurevol/measurevol.html">
<b>Measure Volume and Area</b></a>
- measure surface area and surface-enclosed volume
<li><a href="../ContributedSoftware/minimize/minimize.html">
<b>Minimize Structure</b></a>
- energy-minimize structures
<li><a href="../ContributedSoftware/minrmsplot/minrmsplot.html">
<b>Minrms Plot</b></a>
- examine structural alignments of proteins from
<a href="http://www.cgl.ucsf.edu/Research/minrms/index.html" target="_blank">
MinRMS</a> (obtained separately,
<a href="http://www.cgl.ucsf.edu/Research/minrms/index.html" target="_blank">
available</a> as source code)
<li><a href="../ContributedSoftware/model/model.html"><b>Model Loops</b></a>
- interface to
<a href="http://salilab.org/modeller/index.html" target="_blank">MODELLER</a>
(obtained separately) for generating alternative conformations of peptide
segments already present in a structure
<li><a href="../UsersGuide/modelpanel.html"><b>Model Panel</b></a>
- list and act on models
<li><a href="../ContributedSoftware/morph/morph.html"><b>Morph
Conformations</b></a>
- create a trajectory that morphs between structures
<li><a href="../ContributedSoftware/morphmap/morphmap.html"><b>Morph 
Map</b></a>
- morph between two related sets of volume data
<li><a href="../ContributedSoftware/movemode/movemode.html"><b>Movement
Mouse Mode</b></a>
- move just part of a structure with the mouse
<li><a href="../ContributedSoftware/recorder/recorder.html" target="Content">
<b>Movie Recorder</b></a>
- capture image frames and assemble them into a movie file
<li><a href="../ContributedSoftware/multalignviewer/framemav.html"
target="_top"><b>Multalign Viewer</b></a>
- view sequence alignments, optionally with structures
<li><a href="../ContributedSoftware/multiscale/framemulti.html" target="_top">
<b>Multiscale Models</b></a>
- view macromolecular assemblies at high and low resolution,
generate multimers, and navigate structural hierarchies
<li><a href="../ContributedSoftware/nucleotides/nucleotides.html">
<b>Nucleotides</b></a>
- create special representations of nucleotide bases and sugars
<li><a href="../ContributedSoftware/per-model/per-model.html">
<b>Per-Model&nbsp;Clipping</b></a>
- clip models individually with a plane at any angle
<li><a href="../ContributedSoftware/pipesandplanks/pipesandplanks.html">
<b>PipesAndPlanks</b></a>
- show peptide/protein helices as "pipes" (cylinders) and strands as "planks"
<li><a href="pbpanel.html" target="Content"><b>PseudoBond Panel</b></a>
- list and act on <a href="pbpanel.html#pbdef">pseudobond</a> groups
<li><a href="../ContributedSoftware/pbreader/pbreader.html">
<b>PseudoBond Reader</b></a>
- create <a href="pbpanel.html#pbdef">pseudobonds</a> arbitrarily
<li><a href="../ContributedSoftware/rainbow/rainbow.html" target="Content">
<b>Rainbow</b></a>
- use a range of colors, changing the color per residue, chain, or model
<li><a href="../ContributedSoftware/readstdin/readstdin.html" target="Content">
<b>ReadStdin</b></a>
- allow communication with Chimera through standard input/output
<li><a href="../ContributedSoftware/render/render.html#render">
<b>Render by Attribute</b></a>
- show 
<a href="../ContributedSoftware/defineattrib/defineattrib.html#attribdef">
attribute</a> values of atoms, residues, and models (with color, <i>etc.</i>)
<li><a href="../UsersGuide/reply.html"><b>Reply Log</b></a>
- show informational, warning, and error messages from Chimera
<li><a href="../ContributedSoftware/resprop/resprop.html"><b>ResProp</b></a>
- define amino acid categories
<li><a href="../ContributedSoftware/ribbonstyle/ribbonstyle.html">
<b>Ribbon Style Editor</b></a>
- change ribbon heights/widths (scalings) and cross-sections (styles)
<li><a href="../ContributedSoftware/rotamers/framerot.html"
target="_top"><b>Rotamers</b></a>
- view and evaluate amino acid sidechain rotamers, 
incorporate them into structures
<li><a href="sideview.html#rotation" target="Content"><b>Rotation</b></a>
- control the center of rotation
<li><a href="../ContributedSoftware/scalebar/scalebar.html"
target="Content"><b>Scale Bar</b></a>
- draw a scale bar and associated label
<li><a href="../ContributedSoftware/multalignviewer/framemav.html" 
target="_top"><b>Sequence</b></a>
- show amino acid and/or nucleic acid sequence
<li><a href="sideview.html#shininess"
target="Content"><b>Shininess</b></a>
- adjust shininess and brightness
<li><a href="sideview.html#sideview" target="Content"><b>Side View</b></a>
- scale the view and move clipping planes interactively
<li><a href="../ContributedSoftware/solvate/solvate.html"
target="Content"><b>Solvate</b></a>
- solvate structures using
<a href="http://ambermd.org/#AmberTools" target="_blank">AmberTools</a>
<li><a href="../ContributedSoftware/diagram/diagram.html"
target="Content"><b>Structure Diagram</b></a>
- generate 2D chemical diagrams of small molecules
<li><a href="../ContributedSoftware/surfcapper/surfcapper.html"
target="Content"><b>Surface Capping</b></a>
- cap surfaces where they intersect a clipping plane
<li><a href="../ContributedSoftware/surfcolor/surfcolor.html"
target="Content"><b>Surface Color</b></a>
- color surfaces by volume data values or by distance
from a point, axis, or plane
<li><a href="../ContributedSoftware/surfzone/surfzone.html" 
target="Content"><b>Surface Zone</b></a>
- restrict the display of
<a href="../ContributedSoftware/surfzone/surfzone.html">certain types
of surfaces</a> to a zone around <a href="selection.html">selected</a> atoms
<li><a href="../ContributedSoftware/surfnet/surfnet.html"><b>Surfnet</b></a>
- examine cavities and surface indentations
<li><a href="../ContributedSoftware/idle/task.html"><b>Task Panel</b></a>
- manage jobs started by Chimera
<li><a href="../ContributedSoftware/thermal/thermal.html"><b>Thermal 
Ellipsoids</b></a>
- show anisotropic B-factors
<li><a href="../ContributedSoftware/ensembletile/ensembletile.html">
<b>Tile Structures</b></a>
- arrange models in a plane
<li><a href="../ContributedSoftware/transform/transform.html"><b>Transform
Molecule Coordinates</b></a>
- transform a structure by specified Euler angles and shifts
<li><a href="../ContributedSoftware/undomove/undomove.html"><b>Undo Move</b>
and <b>Redo Move</b></a>
- go back and forth in the history of model rotational/translational positions
<li><a href="../ContributedSoftware/unitcell/unitcell.html">
<b>Unit Cell</b></a>
- generate crystallographic unit cell contents from coordinates
and transformation matrices 
<li><a href="../ContributedSoftware/density/density.html"
target="Content"><b>Values at Atom Positions</b></a>
- map volume data to atom positions and assign values as an
<a href="../ContributedSoftware/defineattrib/defineattrib.html#attribdef">
attribute</a>
<li><a href="../ContributedSoftware/viewdock/framevd.html"
target="_top"><b>ViewDock</b></a>
- view and prioritize docked molecules output by
<a href="http://dock.compbio.ucsf.edu/index.html" target="_blank">DOCK</a>
<li><a href="../ContributedSoftware/voleraser/voleraser.html">
<b>Volume Eraser</b></a>
- interactively zero out parts of volume data
<li><a href="../ContributedSoftware/volumeviewer/gaussian.html"><b>Volume
Filter</b></a>
- smooth or transform volume data
<li><a href="../ContributedSoftware/volumeviewer/volstats.html"><b>Volume
Mean, SD, RMS</b></a>
- calculate statistics for <a href="filetypes.html#volume">volume data</a>
<li><a href="../ContributedSoftware/volseries/volseries.html">
<b>Volume Series</b></a>
- display a series of <a href="filetypes.html#volume">volume data</a> sets
<li><a href="../ContributedSoftware/volumepathtracer/framevolpath.html"
target="_top"><b>Volume Tracer</b></a>
- place markers and trace paths within volume displays
<li><a href="../ContributedSoftware/volumeviewer/framevolumeviewer.html"
target="_top"><b>Volume Viewer</b></a>
- visualize <a href="filetypes.html#volume">volume data</a>
(3D numerical data such as electron density)
<li><a href="dms.html#writedms"><b>Write DMS</b></a>
- save <a href="representation.html#surfaces">molecular surfaces</a>
as <a href="dms.html">DMS files</a>
<li><a href="../ContributedSoftware/addions/writeprmtop.html"><b>Write 
Prmtop</b></a>
- write <a href="http://ambermd.org/index.html" target="_blank"><b>Amber</b></a>
parameter/topology files using
<a href="http://ambermd.org/#AmberTools" target="_blank">AmberTools</a>
</ol>
</body>
</html>
