<!--
--- UCSF Chimera Copyright ---
Copyright (c) 2004-2009 Regents of the University of California.
All rights reserved.  This software provided pursuant to a
license agreement containing restrictions on its disclosure,
duplication and use.  This notice must be embedded in or
attached to all copies, including partial copies, of the
software or any revisions or derivations thereof.
--- UCSF Chimera Copyright ---
-->

<html><head>
<title>Render/Select by Attribute</title>
</head><body>

<a name="renderselect">
The <b>Render/Select by Attribute</b> interface</a> 
performs several functions related to 
<a href="../defineattrib/defineattrib.html#attribdef">attributes</a>:
<ul>
<li><a href="#render">visual rendering</a> of attribute values
<li><a href="#select">selection</a> by attribute values
<li><a href="#saving">saving</a> attribute values to an
<a href="../defineattrib/defineattrib.html#attrfile">assignment file</a>
</ul>

<h2><a name="render">Render by Attribute</a> 
<img src="rendericon.png" alt="Render by Attribute icon"></h2>

<p>
<b>Render by Attribute</b> shows 
<a href="../defineattrib/defineattrib.html#attribdef">attribute</a> values
of atoms, residues, and models:
</p>
<table align="center" cellspacing="2" style="margin:8px 8px 8px 8px">
<tr>
<th align="center"><a href="#colors">Colors</a></th>
<th align="center"><a href="#radii">Radii</a></th>
<th align="center"><a href="#worms">Worms</a></th>
</tr>
<tr><td>
<img src="1bzm-al2co.png" alt="residue mavConservation (AL2CO)">
</td><td>
<img src="1bzm-bfactor.png" alt="alpha-carbon bfactor">
</td><td>
<img src="1bzm-hydrophobicity.png" alt="residue kdHydrophobicity">
</td></tr>
</table>
<p>
There are <a href="../../UsersGuide/extension.html">several ways to start</a>
<b>Render by Attribute</b>, a tool in the <b>Depiction</b> 
and <b>Structure Analysis</b> categories.
</p><p>
<a name="choosemodel">
The first step is to choose one or more models.
</a>
The <b>Models</b> list includes all open molecule models.
Individual models or blocks of models can be chosen with the left mouse
button.  Ctrl-click toggles the status of an individual model.
To choose a block of models without dragging,
click on the first (or last) and then Shift-click 
on the last (or first) in the desired block.
</p><p>
The next step is to choose an
<a href="../defineattrib/defineattrib.html#attribdef">attribute</a>.
Only the attributes of the <a href="#choosemodel">chosen models</a>
will be listed.  For rendering, these may include the
<a name="numattrib">numerical attributes</a>:
<ul>
<li><b>atoms</b>
  <ul>
  <li><b>areaSAS</b> - <a href="../../UsersGuide/representation.html#surfaces">
  solvent-accessible</a> surface area (available when a 
  <a href="../../UsersGuide/representation.html#surfaces">molecular
  surface</a> has been computed)
  <li><b>areaSES</b> - <a href="../../UsersGuide/representation.html#surfaces">
  solvent-excluded</a> surface area (available when a 
  <a href="../../UsersGuide/representation.html#surfaces">molecular
  surface</a> has been computed)
  <li><b>bfactor</b> - B-factor (temperature factor; 
  generally read from
  <a href="../../UsersGuide/tutorials/pdbintro.html#coords">PDB input</a>)
  <li><b>occupancy</b> - occupancy (generally read from
  <a href="../../UsersGuide/tutorials/pdbintro.html#coords">PDB input</a>)
  </ul>
<li><b>residues</b>
  <ul>
  <li><b>areaSAS</b> - <a href="../../UsersGuide/representation.html#surfaces">
  solvent-accessible</a> surface area (available when a 
  <a href="../../UsersGuide/representation.html#surfaces">molecular
  surface</a> has been computed)
  <li><b>areaSES</b> - <a href="../../UsersGuide/representation.html#surfaces">
  solvent-excluded</a> surface area (available when a 
  <a href="../../UsersGuide/representation.html#surfaces">molecular
  surface</a> has been computed)
<li><b>kdHydrophobicity</b> -
<a href="../../UsersGuide/midas/hydrophob.html">Kyte-Doolittle 
hydrophobicity</a> (available only for standard amino acids)
  <li><a href="../multalignviewer/multalignviewer.html#mavPercentConserved"><b>mavPercentConserved</b></a>
- conservation at the associated position of a sequence alignment 
(this and other
<a href="../multalignviewer/multalignviewer.html#mavAttributes">
sequence-derived attributes</a>
are available only for residues associated with an alignment open in
<a href="../multalignviewer/framemav.html" target="_top"><b>Multalign 
Viewer</b></a>)
  <li><b>average</b> - average of the values of the constituent atoms, 
    considering only those with values for that attribute
    <ul>
    <li><i>atom attribute 1</i>
    <li><i>atom attribute 2</i>
    <li> ...
    </ul>
  <li><b>total</b> - total of the values of the constituent atoms
    <ul>
    <li><i>atom attribute 3</i>
    <li><i>atom attribute 4</i>
    <li> ...
    </ul>
  </ul>
<li><b>models</b>
  <ul>
  <li><b>average</b> - average of the values of the consituent residues,
    considering only those with values for that attribute
    <ul>
    <li><i>residue attribute 1</i>
    <li><i>residue attribute 2</i>
    <li> ...
    </ul>
  <li><b>total</b> - total of the values of the consituent residues
    <ul>
    <li><i>residue attribute 3</i>
    <li><i>residue attribute 4</i>
    <li> ...
    </ul>
  </ul>
</ul>
The name of an (atom, residue) attribute controls whether it is automatically
available as an average or total at the next higher level (residue, model).
If <b>area</b>, <b>charge</b>, or <b>volume</b> is a component of the name,
the attribute will be totaled; if not, it will be averaged.  Name components
are indicated with underscores or camel-caps. For example, the components
of <b>my_charge</b> are <b>my</b> and <b>charge</b>; the components of
<b>voronoiVolume</b> are <b>voronoi</b> and <b>volume</b>.
</p><p>
Additional numerical attributes can be:
<ul>
<li> created arbitrarily with
<a href="../defineattrib/defineattrib.html"><b>Define
Attribute</b></a>
<li> created by combining other attributes
(<a href="../calculator/calculator.html"><b>Attribute
Calculator</b></a>)
<li> generated by using certain other Chimera tools (for example, 
<a href="../density/density.html"><b>Values at Atom Positions</b></a>)
</ul>

</p><p>
<a name="refreshing">
If <b>Render/Select by Attribute</b>
was opened before values were assigned for some attribute</a>, 
the new attribute can be listed using <b>Refresh... Menus</b>.
When attribute values have been
changed or parts of the <a href="#choosemodel">chosen models</a> have
been deleted, the display can be updated with <b>Refresh... Values</b>.
The data will refresh automatically, however,
when the set of <a href="#choosemodel">chosen models</a> is changed.
</p><p>
Once an attribute has been chosen, a histogram of the corresponding
values is shown.  Histogram bar heights are logarithmic by default, but
<a name="scaling">can be changed to linear (and back)
with the <b>Scaling</b> menu.</a>  
</p><p>
<a name="thresh">
<i><b>Thresholds</b></i> for defining the interpolation function</a>
(for <a href="#colors"><b>Colors</b></a>,
<a href="#radii"><b>Radii</b></a>,
or <a href="#worms"><b>Worms</b></a>)
are shown as vertical bars on the histogram.
Clicking on a threshold shows its <b>Value</b> in black.
Clicking elsewhere within the histogram shows the <b>Value</b>
(X-coordinate) of the mouseclick in gray.
A threshold can be moved by changing its <b>Value</b>
and then pressing Enter (return) or by dragging it horizontally 
with the left mouse button.  Holding the Shift
key down reduces the speed (mouse sensitivity) of threshold
dragging tenfold, allowing finer control.
Thresholds can be added by Ctrl-clicking
with the left mouse button on the histogram.  Ctrl-clicking on an
existing threshold deletes it.
</p><p>
The rendering can be restricted to the current 
<a href="../../UsersGuide/selection.html">selection</a>
(<b>Restrict OK/Apply...</b>).
<b>OK</b> performs the rendering and dismisses the dialog,
whereas <b>Apply</b> performs the rendering without dismissing the dialog.
<b>Close</b> simply dismisses the dialog. <b>Help</b> opens
this manual page in a browser window.
</p>

<a name="colors">
<h3>Colors</h3></a>
<p>
Atoms and <a href="../../UsersGuide/representation.html#surfaces">surfaces</a>
(molecular, VDW) can be colored by atom, residue, or model
attribute values; ribbons can be colored by residue or model attribute
values.  Besides a <b>Value</b>
(position on the histogram), each <a href="#thresh">threshold</a>
has a <b>Color</b>.
<a name="threshcolor">
The <b>Color</b> applies to the threshold most recently moved or clicked
</a>
and can be adjusted by clicking the adjacent
<a href="../../UsersGuide/coloring.html#colorwell">color well</a>.
</p><p>
The <a href="#thresh">thresholds</a> 
define a function that maps attribute values to colors.
For each atom (residue, model) the attribute value is compared to the 
thresholds on the histogram.
Color is defined by red, green, blue and opacity/transparency components.
The color of the closest threshold at a lower value (to the left) and 
the color of the closest threshold at a higher value (to the right) 
are linearly interpolated.
Items with attribute values less than the leftmost threshold are colored
according to the leftmost threshold, while 
items with attribute values greater than the rightmost threshold are colored
according to the rightmost threshold. 
The <b>Keep opaque</b> options for atom and ribbon coloring indicate that
any transparency in the color definitions should be ignored.
</p><p>
There may be cases where the atoms (residues, models) have 
<a name="novalue">no value</a> for the chosen attribute.
For example, nucleic acid residues do not have 
<a href="../../UsersGuide/midas/hydrophob.html"><b>kdHydrophobicity</b></a>
values, and residues not associated with a position in a sequence alignment 
open in <a href="../multalignviewer/framemav.html" target="_top"><b>Multalign
Viewer</b></a> do not have <b>mavPercentConserved</b> values.
In these cases, the <b>No-value color</b> (also shown in a
<a href="../../UsersGuide/coloring.html#colorwell">color well</a>)
applies.
</p><p>
<b>Create corresponding color key</b> opens the
<a href="../2dlabels/2dlabels.html#colorkey"><b>Color Key</b></a> dialog
and populates it with the current <a href="#thresh">threshold</a> colors 
and values; a color key can then be created interactively with the mouse.
</p><p>
See also:
<a href="../../UsersGuide/midas/rangecolor.html"><b>rangecolor</b></a>,
<a href="../density/density.html"><b>Values at Atom Positions</b></a>,
<a href="../surfcolor/surfcolor.html"><b>Surface Color</b></a>,
<a href="../coulombic/coulombic.html"><b>Coulombic Surface Coloring</b></a>
</p>

<a name="radii">
<h3>Radii</h3></a>
<p>
Atom radii can be assigned according to atom, residue, or model
attribute values.  The <b>Atom style</b> controls which
<a href="../../UsersGuide/drawmode.html">draw mode</a> will be used
for the affected atoms (<b>ball</b> or <b>sphere</b>).  
Ball radius equals atom radius multiplied by a model's 
<a href="../../UsersGuide/moleculeattrib.html#ballscale">ball scale</a> 
factor, whereas sphere radius is the same as atom radius.
</p><p>
Besides a <b>Value</b>
(position on the histogram), each <a href="#thresh">threshold</a>
has an associated <b>Atom radius</b>, expressed in angstroms.
The <b>Atom radius</b> value
corresponds to the threshold most recently moved or clicked
and can be adjusted by entering a different number.
</p><p>
The <a href="#thresh">thresholds</a>
define a function that maps attribute values to atom radii.
For each atom (residue, model) the attribute value is compared to the 
thresholds on the histogram.
The atom radius of the closest threshold at a lower value (to the left) 
and the atom radius of the closest threshold at a higher value (to the right) 
are linearly interpolated.  Atoms (or atoms within residues or models)
with attribute values less than the leftmost threshold are assigned
the atom radius of the leftmost threshold, while
those with attribute values greater than the rightmost threshold 
are assigned the atom radius of the rightmost threshold.
</p><p>
When there are atoms (residues, models) with
<a href="#novalue">no value</a> for the chosen attribute:
if <b>Affect no-value atoms</b> is <b>true</b>,
atoms (or atoms within residues or models) with 
no value will be assigned the <b>No-value radius</b>.
Otherwise, the radii of these atoms will not be changed.
</p><p>
Radii can be returned to their 
<a href="../../UsersGuide/midas/vdwrad.html">default values</a> 
with the command
<a href="../../UsersGuide/midas/vdwdefine.html"><b>~vdwdefine</b></a>.
</p>

<a name="worms">
<h3>Worms</h3></a>
<p>
Worm radii can be assigned according to residue or model attribute values.
Worms are ribbons with a special
<a href="../ribbonstyle/ribbonstyle.html#scaling">scaling</a>
that depends on attribute values instead of secondary structure.  
Rendering as worms displays a ribbon for the affected residues in the
<a href="../ribbonstyle/ribbonstyle.html#xsection">style</a> named
<a href="../ribbonstyle/ribbonstyle.html#builtinstyles"><b>rounded</b></a>
with the special
<a href="../ribbonstyle/ribbonstyle.html#scaling">scaling</a>.
Like other ribbons, worms 
can be shown only for amino acid and nucleic acid residues
bonded to at least one other residue.
The <b>Worm style</b> controls whether
the radius is smoothed continuously along the ribbon (<b>smooth</b>)
or changes abruptly for each residue (<b>segmented</b>).
To obtain standard, secondary-structure-dependent
ribbons after rendering as worms, it is necessary to render
again with the <b>Worm style</b> set to <b>non-worm</b>.
</p><p>
Besides a <b>Value</b>
(position on the histogram), each <a href="#thresh">threshold</a>
has an associated <b>Worm radius</b>, expressed in angstroms.
The <b>Worm radius</b> value
corresponds to the threshold most recently moved or clicked
and can be adjusted by entering a different number.
</p><p>
The <a href="#thresh">thresholds</a>
define a function that maps attribute values to worm radii.
For each residue (or model) the attribute value is compared to the 
thresholds on the histogram.
The worm radius of the closest threshold at a lower value (to the left)
and the worm radius of the closest threshold at a higher value (to the right)
are linearly interpolated.  Residues (or residues within models)
with attribute values less than the leftmost threshold are assigned
the worm radius of the leftmost threshold, while
those with attribute values greater than the rightmost threshold 
are assigned the worm radius of the rightmost threshold.
</p><p>
When there are residues (or models) with
<a href="#novalue">no value</a> for the chosen attribute:
if <b>Affect no-value residues</b> is <b>true</b>,
residues (or residues within models) with no value
will be shown as worms with the <b>No-value worm radius</b>, 
unless the <b>Worm style</b>
is <b>non-worm</b>.  Rendering with the <b>non-worm</b> style
restores the standard ribbon for all affected residues.
</p>

<h2><a name="select">Select by Attribute</a></h2> 
<p>
<b>Select by Attribute</b> allows
<a href="../../UsersGuide/selection.html">selection</a>
of atoms, residues and models by their 
<a href="../defineattrib/defineattrib.html#attribdef">attribute</a>
values.
It can be opened by choosing <b>By Attribute Value...</b> from the Chimera
<a href="../../UsersGuide/menu.html#menuselect"><b>Select</b> menu</a> or
<a href="../multalignviewer/multalignviewer.html#mavmenu-structure">
<b>Structure... Select by Conservation</b></a> from the menu in
in <a href="../multalignviewer/framemav.html" target="_top"> 
<b>Multalign Viewer</b></a>.
</p><p>
One must <a href="#choosemodel">choose one or more models</a>
and then one of the
<a href="../defineattrib/defineattrib.html#attribdef">attributes</a>
available for the chosen models.
These may include the <a href="#numattrib">numerical attributes</a>
available for <a href="#render">rendering</a> plus the following 
string-valued and boolean attributes:
<ul>
<li><b>atoms</b>
  <ul>
  <li><b>display</b> - whether display is enabled at the atom level
(see <a href="../../UsersGuide/hierarchy.html#display">display hierarchy</a>)
  <li><b>idatmType</b> - <a href="../../UsersGuide/idatm.html">atom type</a>
  <li><b>surfaceCategory</b> - the name of the category to which the atom
  has been assigned 
  <a href="../../UsersGuide/midas/surface.html#surfcats">automatically</a>
  or manually using 
  <a href="../../UsersGuide/midas/msms.html"><b>surfcat</b></a>
  </ul>
<li><b>residues</b>
  <ul>
  <li><b>isHelix</b> - whether the residue is in a helix (true is
   only possible for amino acids)
  <li><b>isHet</b> - whether the residue is in HETATM records 
   in the input PDB file
  <li><b>isStrand</b> - whether the residue is in a beta strand (true is
   only possible for amino acids)
  <li><b>isTurn</b> - whether the residue is in a turn according to PDB
   TURN records (true is only possible for amino acids)
  <li><b>type</b> - residue name 
  </ul>
</ul>
Additional string-valued and boolean attributes can be:
<ul>
<li> created arbitrarily with
<a href="../defineattrib/defineattrib.html"><b>Define
Attribute</b></a>
<li> generated by using certain other Chimera tools (for example,
<a href="../nucleotides/nucleotides.html"><b>Nucleotides</b></a>)
</ul>
</p><p>
Once an attribute has been chosen, a <a href="#selhist">histogram</a>
(if numerical), <a href="#sellist">list</a> (if character strings),
or <a href="#selbool">set of checkbox options</a> (if boolean)
of the corresponding values is shown.  
The data can be <a href="#refreshing">refreshed</a> 
after new attributes have been added, attribute values have been changed,
or parts of the <a href="#choosemodel">chosen models</a> have been deleted.
</p>
<ul>
<li><a name="selhist">
Within a histogram</a>,
the bar <a href="#scaling">scaling</a> can be logarithmic or linear.
Two <b><i>markers</i></b> are shown as vertical bars.
Clicking on a marker shows its <b>Value</b> in black.
Clicking elsewhere within the histogram shows the <b>Value</b>
(X-coordinate) of the mouseclick in gray.
A marker can be moved by changing its <b>Value</b>
and then pressing Enter (return) or by dragging it horizontally
with the left mouse button.  Holding the Shift
key down reduces the speed (mouse sensitivity) of marker
dragging tenfold, allowing finer control.
The mutually exclusive options are to <b>Select</b>:
  <ul>
  <li><b>between markers (inclusive)</b>
  <li><b>outside markers</b>
  <li><b>no value</b> - applies to atoms (residues, models) 
   with <a href="#novalue">no value</a> for the chosen attribute
   (only listed when present)
  </ul>
<li><a name="sellist">
Within a list of character strings</a>, 
simply clicking an entry toggles its status between will/will not be selected.
<br>
<li><a name="selbool">
For boolean attributes</a>, the mutually exclusive options are:
  <ul>
  <li><b>False</b>
  <li><b>True</b>
  <li><b>No value</b> - applies to atoms (residues, models) 
   with <a href="#novalue">no value</a> for the chosen attribute
   (only listed when present)
  </ul>
</ul>
<p>
<b>OK</b> performs the selection and dismisses the dialog,
whereas <b>Apply</b> performs the selection without dismissing the dialog.
<b>Close</b> simply dismisses the dialog. <b>Help</b> opens
this manual page in a browser window.
</p>

<h3><a name="saving">Saving Attributes</a></h3>
<p>
An attribute and its associated values can be saved to an
<a href="../defineattrib/defineattrib.html#attrfile">attribute assignment
file</a> by choosing <b>File... Save Attributes</b> from the menu of
<a href="#renderselect"><b>Render/Select by Attribute</b></a>.
</p><p>
The attribute to save, its level (<b>atoms</b>, <b>residues</b>, or
<b>models</b>), and the model(s) to include
should be chosen from the menus in the resulting
<a href="../../UsersGuide/opensave.html">save dialog</a>.
Additional options:
</p>
<ul>
<li><b>Restrict save to current selection, if any</b> - 
when a <a href="../../UsersGuide/selection.html">selection</a> exists,
whether to write out only the attribute values of the
<a href="../../UsersGuide/selection.html">selected</a>
items (atoms, residues, models) within the set of chosen models 
<li><b>Include model numbers in output</b> - whether 
<a href="../../UsersGuide/midas/atom_spec.html#basic">atom 
specification strings</a> in the output
<a href="../defineattrib/defineattrib.html#attrfile">assignment file</a> 
should include model numbers.  
Omitting the model number allows later use of the file to assign
attribute values when the model has been opened with a different number.
</ul>
<hr>
<address>UCSF Computer Graphics Laboratory / January 2009</address>
</body></html>
