<!--
--- UCSF Chimera Copyright ---
Copyright (c) 2002-2009 Regents of the University of California.
All rights reserved.  This software provided pursuant to a
license agreement containing restrictions on its disclosure,
duplication and use.  This notice must be embedded in or
attached to all copies, including partial copies, of the
software or any revisions or derivations thereof.
--- UCSF Chimera Copyright ---
-->

<html>
<head>
<TITLE>Aligned FASTA Format</TITLE>
</head>
<body>
<h3>Aligned FASTA Format</h3>
<p>
<a href="framemav.html" target="_top"><b>Multalign Viewer</b></a>
displays sequence alignments and single sequences.
</p><p>
Sequence alignments can be read/written in
<b><i>aligned FASTA format</i></b>, a simple variant of
<a href="http://www.ncbi.nlm.nih.gov/BLAST/fasta.html" 
target="_blank">standard FASTA format</a>.
In an aligned FASTA file, all of the sequences are made the same length
by including characters for leading, trailing, and gap positions.
<a href="framemav.html" target="_top"><b>Multalign Viewer</b></a>
tolerates any non-alphanumeric characters in these positions.  
In <a href="apoex.fa">apoex.fa</a>, for example,
leading and trailing positions are indicated by tildes (~)
and gap positions are indicated by periods (.).
</p><p>
Single sequences can be read/written in
<a href="http://www.ncbi.nlm.nih.gov/BLAST/fasta.html" 
target="_blank">standard FASTA format</a>, but
a file that contains multiple unaligned sequences 
(sequences of different lengths) will not be accepted.
</p>
<hr>
<address>UCSF Computer Graphics Laboratory / November 2009</address>
</body>
</html>
