<!--
--- UCSF Chimera Copyright ---
Copyright (c) 2002-2009 Regents of the University of California.
All rights reserved.  This software provided pursuant to a
license agreement containing restrictions on its disclosure,
duplication and use.  This notice must be embedded in or
attached to all copies, including partial copies, of the
software or any revisions or derivations thereof.
--- UCSF Chimera Copyright ---
-->

<html>
<head>
<TITLE>Aligned NBRF/PIR Format</TITLE>
</head>
<body>
<h3>Aligned NBRF/PIR Format</h3>
<p>
<a href="framemav.html" target="_top"><b>Multalign Viewer</b></a>
displays sequence alignments and single sequences.
</p><p>
Sequence alignments can be read/written in
<b><i>aligned NBRF/PIR format</i></b>, a simple variant of
<a href="http://www.bioinformatics.nl/tools/crab_pir.html" 
target="_blank">standard NBRF/PIR format</a>.
In an aligned NBRF/PIR file, all of the sequences are made the same length
by including characters for leading, trailing, and gap positions.
<a href="framemav.html" target="_top"><b>Multalign Viewer</b></a>
tolerates any non-alphanumeric characters except asterisk (*)
in these positions.  In <a href="rhod.pir">rhod.pir</a>,
for example, dashes (-) are used.
The second line for each sequence is required, but may be blank;
if it is omitted, the first line of sequence will be interpreted
as comments and left out of the sequence.
Spaces within sequences are ignored.
</p><p>
Single sequences can be read/written in
<a href="http://www.bioinformatics.nl/tools/crab_pir.html" 
target="_blank">standard NBRF/PIR format</a>, but
a file that contains multiple unaligned sequences
(sequences of different lengths) will not be accepted.
</p>
<hr>
<address>UCSF Computer Graphics Laboratory / November 2009</address>
</body>
</html>
